{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "353c737e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import time\n",
    "from joblib import load\n",
    "import requests\n",
    "\n",
    "from sklearn.model_selection import train_test_split, cross_validate, cross_val_score\n",
    "from sklearn.linear_model import SGDClassifier\n",
    "from sklearn.preprocessing import MinMaxScaler"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48554cdc",
   "metadata": {},
   "source": [
    "# Load Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "811b90c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/jfschulz_notebooks'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6cc0bb7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Case</th>\n",
       "      <th>years_to_birth</th>\n",
       "      <th>gender</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>radiation_therapy</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Mutation.Count</th>\n",
       "      <th>Percent.aneuploidy</th>\n",
       "      <th>...</th>\n",
       "      <th>p27_p</th>\n",
       "      <th>p27_pT157_p</th>\n",
       "      <th>p27_pT198_p</th>\n",
       "      <th>p38_pT180_Y182_p</th>\n",
       "      <th>p53_p</th>\n",
       "      <th>p62-LCK-ligand_p</th>\n",
       "      <th>p70S6K_p</th>\n",
       "      <th>p70S6K_pT389_p</th>\n",
       "      <th>p90RSK_p</th>\n",
       "      <th>p90RSK_pT359_S363_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-CS-4938</td>\n",
       "      <td>31</td>\n",
       "      <td>female</td>\n",
       "      <td>astrocytoma</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>no</td>\n",
       "      <td>G2</td>\n",
       "      <td>15</td>\n",
       "      <td>0.069412</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.425127</td>\n",
       "      <td>-0.033398</td>\n",
       "      <td>0.289192</td>\n",
       "      <td>1.060163</td>\n",
       "      <td>-0.407456</td>\n",
       "      <td>-0.470354</td>\n",
       "      <td>-0.107559</td>\n",
       "      <td>-0.060441</td>\n",
       "      <td>-0.053104</td>\n",
       "      <td>-0.011132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-CS-6665</td>\n",
       "      <td>51</td>\n",
       "      <td>female</td>\n",
       "      <td>astrocytoma</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>yes</td>\n",
       "      <td>G3</td>\n",
       "      <td>75</td>\n",
       "      <td>0.524814</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076536</td>\n",
       "      <td>0.011809</td>\n",
       "      <td>-0.047973</td>\n",
       "      <td>-0.054275</td>\n",
       "      <td>-0.244020</td>\n",
       "      <td>0.106282</td>\n",
       "      <td>-0.034369</td>\n",
       "      <td>0.585072</td>\n",
       "      <td>0.431720</td>\n",
       "      <td>-0.201412</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 186 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Case  years_to_birth  gender histological_type   race  \\\n",
       "0  TCGA-CS-4938              31  female       astrocytoma  white   \n",
       "1  TCGA-CS-6665              51  female       astrocytoma  white   \n",
       "\n",
       "                ethnicity radiation_therapy Grade  Mutation.Count  \\\n",
       "0  not hispanic or latino                no    G2              15   \n",
       "1  not hispanic or latino               yes    G3              75   \n",
       "\n",
       "   Percent.aneuploidy  ...     p27_p  p27_pT157_p  p27_pT198_p  \\\n",
       "0            0.069412  ... -0.425127    -0.033398     0.289192   \n",
       "1            0.524814  ...  0.076536     0.011809    -0.047973   \n",
       "\n",
       "   p38_pT180_Y182_p     p53_p  p62-LCK-ligand_p  p70S6K_p  p70S6K_pT389_p  \\\n",
       "0          1.060163 -0.407456         -0.470354 -0.107559       -0.060441   \n",
       "1         -0.054275 -0.244020          0.106282 -0.034369        0.585072   \n",
       "\n",
       "   p90RSK_p  p90RSK_pT359_S363_p  \n",
       "0 -0.053104            -0.011132  \n",
       "1  0.431720            -0.201412  \n",
       "\n",
       "[2 rows x 186 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins.csv\" \n",
    "df = pd.read_csv(f\"{path}/{file}\", header=0)\n",
    "#df = pd.read_csv(path)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "70bb2a5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['astrocytoma', 'oligoastrocytoma', 'oligodendroglioma'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df.info(verbose=True, show_counts=True)\n",
    "df[\"histological_type\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e49f046",
   "metadata": {},
   "source": [
    "# Define Features and Target"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77df6136-8ab7-4299-9fd8-3ab6b7fa6d04",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a51a2b9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define target and features\n",
    "cols_to_drop = ['Case', 'years_to_birth', 'gender', 'histological_type', 'race',\n",
    "       'ethnicity', 'radiation_therapy', 'Grade', 'Mutation.Count',\n",
    "       'Percent.aneuploidy', 'IDH.status', 'outcome']\n",
    "X = df.drop(cols_to_drop, axis = 1)\n",
    "y = df[\"outcome\"]\n",
    "\n",
    "#X.info(verbose=True, show_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c7864bf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create train and test data\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a3a27b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save X_test data set to use for API testing:\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-test-with-identifier.csv\" \n",
    "\n",
    "X_test.to_csv(path_or_buf=f\"{path}/{file}\", header=True, index=True, index_label=\"Identifier\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "446c72bc-f0a3-455a-9e5b-69ed65dc0243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save one row of df data set to use for API testing: row 85 --> outcome = 1\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-test-with-identifier-outcome1.csv\" \n",
    "cols_to_drop = ['Case', 'years_to_birth', 'gender', 'histological_type', 'race',\n",
    "       'ethnicity', 'radiation_therapy', 'Grade', 'Mutation.Count',\n",
    "       'Percent.aneuploidy', 'IDH.status', 'outcome']\n",
    "\n",
    "tmp1 = df[85:86].drop(cols_to_drop, axis = 1)\n",
    "tmp1.to_csv(path_or_buf=f\"{path}/{file}\", header=True, index=True, index_label=\"Identifier\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ebebe7e6-7882-4883-b5d4-59d2af1aaedd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save one row of df data set to use for API testing: row 274 --> outcome = 0\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-test-with-identifier-outcome0.csv\" \n",
    "cols_to_drop = ['Case', 'years_to_birth', 'gender', 'histological_type', 'race',\n",
    "       'ethnicity', 'radiation_therapy', 'Grade', 'Mutation.Count',\n",
    "       'Percent.aneuploidy', 'IDH.status', 'outcome']\n",
    "\n",
    "tmp1 = df[274:275].drop(cols_to_drop, axis = 1)\n",
    "tmp1.to_csv(path_or_buf=f\"{path}/{file}\", header=True, index=True, index_label=\"Identifier\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7b066344-207e-44cd-9232-fe0530330e95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "274    0\n",
       "Name: outcome, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[274:275][\"outcome\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c1f339e-6d26-4b57-9f2c-700fe03878d6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Save test data that has all columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "44b85ffb-8ab5-402a-9295-41445191e357",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define target and features\n",
    "X = df.drop(\"outcome\", axis = 1)\n",
    "y = df[\"outcome\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "28c06adc-3fce-44cb-ab53-b9b4da91a357",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "31e316bf-f9c0-4410-ade8-918745de571b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save test data set to use for API testing\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-all-cols-test-with-identifier.csv\" \n",
    "X_test.to_csv(path_or_buf=f\"{path}/{file}\", header=True, index=True, index_label=\"Identifier\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e830919-806d-4e01-aceb-e59cb981618a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save one row of df data set to use for API testing: row 85 --> outcome = 1\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-all-cols-test-with-identifier-outcome1.csv\" \n",
    "cols_to_drop = ['outcome']\n",
    "\n",
    "tmp1 = df[85:86].drop(cols_to_drop, axis = 1)\n",
    "tmp1.to_csv(path_or_buf=f\"{path}/{file}\", header=True, index=True, index_label=\"Identifier\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ec8d23c5-e17c-4a8a-9dc9-4164f4908e6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save one row of df data set to use for API testing: row 85 --> outcome = 1\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-all-cols-test-with-identifier-outcome1.csv\" \n",
    "cols_to_drop = ['outcome']\n",
    "\n",
    "tmp1 = df[85:86].drop(cols_to_drop, axis = 1)\n",
    "tmp1.to_csv(path_or_buf=f\"{path}/{file}\", header=True, index=True, index_label=\"Identifier\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "04cfa0f5-5859-4fc5-9f4c-0526a2b1f59b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save one row of df data set to use for API testing: row 274 --> outcome = 0\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-all-cols-test-with-identifier-outcome0.csv\" \n",
    "cols_to_drop = ['outcome']\n",
    "\n",
    "tmp1 = df[274:275].drop(cols_to_drop, axis = 1)\n",
    "tmp1.to_csv(path_or_buf=f\"{path}/{file}\", header=True, index=True, index_label=\"Identifier\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "614edc0e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Code to create feature table based on uploaded csv file for api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "76112572",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to create \n",
    "tmp = pd.read_csv(f\"{path}/{file}\", header=0)\n",
    "tmp.head(3)\n",
    "\n",
    "if 'Identifier' in tmp.columns:\n",
    "    X_pred = tmp.drop([\"Identifier\"], axis = 1)\n",
    "else:\n",
    "    X_pred = tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ef5ef7d",
   "metadata": {},
   "source": [
    "# Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "290ac6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scale\n",
    "min_max_scaler = MinMaxScaler()\n",
    "min_max_scaler.fit(X_train)\n",
    "X_train_proc = min_max_scaler.transform(X_train)\n",
    "X_test_proc = min_max_scaler.transform(X_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fe29215",
   "metadata": {},
   "source": [
    "## Functions to save and load scaler for api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b4116aa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save scaler\n",
    "\n",
    "from joblib import dump, load\n",
    "import time\n",
    "\n",
    "\n",
    "def save_scaler(scaler_to_save = None, \n",
    "               scaler_type = None, \n",
    "               path_to_save = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers\"\n",
    "              ):\n",
    "    \"\"\"\n",
    "    Persist trained model locally on the hard drive at f\"{path_to_save/scaler_type/f\"{timestamp}.joblib\"\n",
    "    \"\"\"\n",
    "\n",
    "    timestamp = time.strftime(\"%Y%m%d-%H%M%S\")\n",
    "    \n",
    "    # Save scaler locally\n",
    "    scaler_path_file = os.path.join(f\"{path_to_save}/{scaler_type}_{timestamp}.joblib\")\n",
    "   \n",
    "    dump(scaler_to_save, scaler_path_file) \n",
    "\n",
    "    print(f\"âœ… Scaler saved locally at {scaler_path_file}\")\n",
    "    \n",
    "# Load scaler\n",
    "def load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "               file = 'MinMax_20240306-102844.joblib'\n",
    "              ):\n",
    "    path_file = f\"{path}/{file}\"\n",
    "\n",
    "    scaler = load(path_file) \n",
    "    return scaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9286a811",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "âœ… Scaler saved locally at /home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers/MinMax_20240307-101324.joblib\n"
     ]
    }
   ],
   "source": [
    "save_scaler(scaler_to_save= min_max_scaler,\n",
    "           scaler_type=\"MinMax\"\n",
    "           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a4a2c4d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator MinMaxScaler from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: black;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"â–¸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"â–¾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 1ex;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>MinMaxScaler()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;MinMaxScaler<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.4/modules/generated/sklearn.preprocessing.MinMaxScaler.html\">?<span>Documentation for MinMaxScaler</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>MinMaxScaler()</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "MinMaxScaler()"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "load_scaler(file =\"MinMax_20240306-102844.joblib\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01198d65",
   "metadata": {},
   "source": [
    "# Build Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "38996190",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize model\n",
    "model = SGDClassifier(loss = \"log_loss\")\n",
    "\n",
    "# Cross validte model performance\n",
    "np.mean(cross_val_score(model, X_train_proc, y_train, cv=5))\n",
    "\n",
    "# Train model\n",
    "history = model.fit(X_train_proc, y_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05b44c6e",
   "metadata": {},
   "source": [
    "## Functions to save and load model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "354bc4d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save model\n",
    "\n",
    "from joblib import dump, load\n",
    "import time\n",
    "\n",
    "\n",
    "def save_model(model_to_save = None, \n",
    "               model_type = None, \n",
    "               path_to_save = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_models\"\n",
    "              ):\n",
    "    \"\"\"\n",
    "    Persist trained model locally on the hard drive at f\"{path_to_save/model_type/f\"{timestamp}.joblib\"\n",
    "    \"\"\"\n",
    "\n",
    "    timestamp = time.strftime(\"%Y%m%d-%H%M%S\")\n",
    "    \n",
    "    # Save model locally\n",
    "    model_path_file = os.path.join(f\"{path_to_save}/{model_type}_{timestamp}.joblib\")\n",
    "   \n",
    "    dump(model_to_save, model_path_file) \n",
    "\n",
    "    print(f\"âœ… Model saved locally at {model_path_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1b418a50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "âœ… Model saved locally at /home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_models/SGDclassifier_20240307-101327.joblib\n"
     ]
    }
   ],
   "source": [
    "save_model(model_to_save = model, model_type = \"SGDclassifier\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "622839b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load model\n",
    "def load_model(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_models',\n",
    "               file = 'SGDclassifier_20240305-135742.joblib'\n",
    "              ):\n",
    "    path_file = f\"{path}/{file}\"\n",
    "\n",
    "    model = load(path_file) \n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d38c8410",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator MinMaxScaler from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "model3 = load_model()\n",
    "scaler = load_scaler()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ff01f136-77e2-49b0-88de-4bb13d54a67c",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_test.shape\n",
    "X_test_proc = scaler.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "686d52fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(62, 174)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_test_proc.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b414897b-d061-4b94-9a98-c9856f5684b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(62,)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model3.predict(X_test_proc).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ac32b79c-c887-46be-91c5-d0e91ec82642",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Outcome\n",
       "0         0\n",
       "1         0\n",
       "2         0\n",
       "3         0\n",
       "4         0\n",
       "..      ...\n",
       "57        0\n",
       "58        0\n",
       "59        0\n",
       "60        1\n",
       "61        0\n",
       "\n",
       "[62 rows x 1 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(model.predict(X_test_proc), columns=[\"Outcome\"], dtype = int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4cad610f-83ee-4f19-9030-e49f17dfbec1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Probability_0</th>\n",
       "      <th>Probability_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.539120e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.999490</td>\n",
       "      <td>5.100199e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.376320e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.817148</td>\n",
       "      <td>1.828519e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.463690e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>0.999639</td>\n",
       "      <td>3.609115e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.839377e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.390907e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>0.000056</td>\n",
       "      <td>9.999436e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>0.999993</td>\n",
       "      <td>7.415450e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Probability_0  Probability_1\n",
       "0        1.000000   1.539120e-09\n",
       "1        0.999490   5.100199e-04\n",
       "2        1.000000   4.376320e-10\n",
       "3        0.817148   1.828519e-01\n",
       "4        1.000000   3.463690e-08\n",
       "..            ...            ...\n",
       "57       0.999639   3.609115e-04\n",
       "58       1.000000   4.839377e-10\n",
       "59       1.000000   4.390907e-17\n",
       "60       0.000056   9.999436e-01\n",
       "61       0.999993   7.415450e-06\n",
       "\n",
       "[62 rows x 2 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(model.predict_proba(X_test_proc), columns=[\"Probability_0\", \"Probability_1\"], dtype = float)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ac64fef6-cbcd-4c73-bbff-1afff23a9701",
   "metadata": {},
   "outputs": [],
   "source": [
    "outcome = pd.DataFrame(model.predict(X_test_proc), columns=[\"Outcome\"], dtype = int)\n",
    "prob = pd.DataFrame(model.predict_proba(X_test_proc), columns=[\"Probability_0\", \"Probability_1\"], dtype = float)\n",
    "\n",
    "    # Merge results into dataframe\n",
    "result = pd.merge(prob,outcome, left_index=True, right_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "85ceb875-a5cb-4ad1-8bde-4c16d4aac40a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax. Perhaps you forgot a comma? (3599829651.py, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[28], line 3\u001b[0;36m\u001b[0m\n\u001b[0;31m    \"Probability_0\": result[\"Probability_0\"].tolist()\u001b[0m\n\u001b[0m                     ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax. Perhaps you forgot a comma?\n"
     ]
    }
   ],
   "source": [
    "{\n",
    "    \"Outcome\": result[\"Outcome\"].tolist,\n",
    "    \"Probability_0\": result[\"Probability_0\"].tolist()\n",
    "    \"Probability_1\": result[\"Probability_1\"].tolist()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a255cfb3-8196-4758-bec0-5ee3e8526e3f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: [0.9999999984608803, 1.539119684558466e-09, 0.0],\n",
       " 1: [0.9994899801104767, 0.0005100198895232571, 0.0],\n",
       " 2: [0.999999999562368, 4.376320422855559e-10, 0.0],\n",
       " 3: [0.8171480614345283, 0.18285193856547163, 0.0],\n",
       " 4: [0.9999999653631049, 3.4636895143770705e-08, 0.0],\n",
       " 5: [0.0007739546753290361, 0.999226045324671, 1.0],\n",
       " 6: [0.9999999999585548, 4.1445134143151145e-11, 0.0],\n",
       " 7: [1.0, 4.0370873986584997e-19, 0.0],\n",
       " 8: [4.553750709845872e-10, 0.9999999995446249, 1.0],\n",
       " 9: [0.9999999999999996, 4.020988942126253e-16, 0.0],\n",
       " 10: [0.999999999963069, 3.693097321429898e-11, 0.0],\n",
       " 11: [0.9999999999943312, 5.668799447208983e-12, 0.0],\n",
       " 12: [1.2311417829646842e-06, 0.999998768858217, 1.0],\n",
       " 13: [3.4369593193694925e-07, 0.9999996563040681, 1.0],\n",
       " 14: [0.9999999999999998, 1.6949150987879817e-16, 0.0],\n",
       " 15: [0.9999998829213026, 1.1707869740514118e-07, 0.0],\n",
       " 16: [0.9999999999999384, 6.162935390664951e-14, 0.0],\n",
       " 17: [0.9999999999999486, 5.1415690101531494e-14, 0.0],\n",
       " 18: [0.9999999999999934, 6.496163497566479e-15, 0.0],\n",
       " 19: [1.0, 4.649990465397418e-17, 0.0],\n",
       " 20: [0.9749428327795151, 0.025057167220484934, 0.0],\n",
       " 21: [1.0, 4.207095162102435e-32, 0.0],\n",
       " 22: [0.9999983887675574, 1.6112324425322968e-06, 0.0],\n",
       " 23: [1.0, 8.7202051259926e-27, 0.0],\n",
       " 24: [0.999807509503613, 0.00019249049638700772, 0.0],\n",
       " 25: [1.0, 4.9583420256542995e-25, 0.0],\n",
       " 26: [3.765663780797013e-09, 0.9999999962343362, 1.0],\n",
       " 27: [1.874915500632568e-08, 0.999999981250845, 1.0],\n",
       " 28: [0.9999999999999988, 1.2219284928907165e-15, 0.0],\n",
       " 29: [1.0, 6.184013669984854e-23, 0.0],\n",
       " 30: [0.0004113336656963451, 0.9995886663343037, 1.0],\n",
       " 31: [0.0074660241395566596, 0.9925339758604433, 1.0],\n",
       " 32: [6.543180992135689e-06, 0.9999934568190079, 1.0],\n",
       " 33: [0.038904064031368035, 0.961095935968632, 1.0],\n",
       " 34: [0.9998536726764836, 0.00014632732351639295, 0.0],\n",
       " 35: [1.0, 9.069389206982804e-23, 0.0],\n",
       " 36: [0.9994266428154642, 0.0005733571845358222, 0.0],\n",
       " 37: [0.9999999999999991, 8.379398795588018e-16, 0.0],\n",
       " 38: [2.7475457220482724e-07, 0.9999997252454278, 1.0],\n",
       " 39: [0.9999999466156622, 5.338433784648896e-08, 0.0],\n",
       " 40: [1.0, 2.112708387802795e-20, 0.0],\n",
       " 41: [0.9999999999711282, 2.8871827474897244e-11, 0.0],\n",
       " 42: [0.7689655477774533, 0.23103445222254673, 0.0],\n",
       " 43: [9.193021378361976e-06, 0.9999908069786216, 1.0],\n",
       " 44: [0.9999999795527712, 2.0447228758124006e-08, 0.0],\n",
       " 45: [0.9999988686324105, 1.131367589481473e-06, 0.0],\n",
       " 46: [1.668847382507721e-08, 0.9999999833115262, 1.0],\n",
       " 47: [4.767883318712585e-06, 0.9999952321166813, 1.0],\n",
       " 48: [0.00022026365498239997, 0.9997797363450176, 1.0],\n",
       " 49: [1.0, 4.0463989996536446e-18, 0.0],\n",
       " 50: [0.9999999999647934, 3.520666387755529e-11, 0.0],\n",
       " 51: [2.3122497250227525e-09, 0.9999999976877503, 1.0],\n",
       " 52: [0.9999999999999982, 1.81808103201896e-15, 0.0],\n",
       " 53: [0.9999839532158366, 1.6046784163493662e-05, 0.0],\n",
       " 54: [3.1621067811959946e-06, 0.9999968378932188, 1.0],\n",
       " 55: [0.3671435570115259, 0.6328564429884741, 1.0],\n",
       " 56: [0.9999999999999557, 4.433931597162451e-14, 0.0],\n",
       " 57: [0.9996390885352595, 0.0003609114647404975, 0.0],\n",
       " 58: [0.9999999995160623, 4.839377130106508e-10, 0.0],\n",
       " 59: [1.0, 4.39090654755336e-17, 0.0],\n",
       " 60: [5.6365619846454607e-05, 0.9999436343801535, 1.0],\n",
       " 61: [0.9999925845496435, 7.4154503565701695e-06, 0.0]}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{k: v.tolist() for k, v in result.iterrows()}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7edd78ac-e211-4e33-8c57-cf951d20ff2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_dict = result.to_dict('series')\n",
    "type(result_dict[\"Probability_0\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fe2d189b-da77-4734-83da-4db09f67111d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "list"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(list(model.predict(X_test_proc)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1b1a0766",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predict test data\n",
    "outcome = pd.DataFrame(model.predict(X_test_proc), columns=[\"Outcome\"])\n",
    "prob = pd.DataFrame(model.predict_proba(X_test_proc), columns=[\"Probability_0\", \"Probability_1\"])\n",
    "\n",
    "# Merge results into dataframe\n",
    "result = pd.merge(prob,outcome, left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84db8be2-83f7-436a-8c9c-6eaa57693c30",
   "metadata": {},
   "source": [
    "## SVC model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "8c147352-0ad0-4839-9fde-f961ba782879",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from fastapi import FastAPI\n",
    "from fastapi.responses import JSONResponse\n",
    "from fastapi.middleware.cors import CORSMiddleware\n",
    "from fastapi import FastAPI, File, UploadFile\n",
    "from io import BytesIO, StringIO\n",
    "import json\n",
    "\n",
    "from sklearn.model_selection import train_test_split, cross_validate, cross_val_score\n",
    "from sklearn.linear_model import SGDClassifier\n",
    "from sklearn.preprocessing import MinMaxScaler\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "532f8e2b-dfd2-4967-acf3-dbf49352b3eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_data (df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Clean raw data:\n",
    "    - remove duplicates\n",
    "    - remove NaNs\n",
    "    \"\"\"\n",
    "\n",
    "    df = df.drop_duplicates()\n",
    "    df = df.dropna(how='any', axis=0)\n",
    "\n",
    "    #clinical_data= ['Case', 'years_to_birth', 'gender', 'histological_type', 'race', 'ethnicity', 'radiation_therapy', 'Grade', 'Mutation.Count', 'Percent.aneuploidy', 'IDH.status', 'outcome' ]\n",
    "\n",
    "    print(\"âœ… Data cleaned\")\n",
    "    return df\n",
    "\n",
    "# Load scaler\n",
    "def load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "               file = 'scaler.gz'\n",
    "              ):\n",
    "    \"\"\"\n",
    "    Loads a locally saved and fitted scaler from the given path and file name.\n",
    "    \"\"\"\n",
    "    path_file = f\"{path}/{file}\"\n",
    "\n",
    "    scaler = load(path_file)\n",
    "    return scaler\n",
    "\n",
    "\n",
    "def preprocess_input(data) -> np.array:\n",
    "    \"\"\"\n",
    "    - process data (proteins, age and gender)\n",
    "    - load fitted scaler and pipline for preprocessing\n",
    "    - return processed data\n",
    "    \"\"\"\n",
    "    print(\" Preprocess input proteins, age and gender... \")\n",
    "\n",
    "    # Clean data\n",
    "    data_clean= clean_data(data)\n",
    "\n",
    "    # define X\n",
    "    X = data_clean.drop(['Case', 'histological_type', 'race', 'ethnicity', 'radiation_therapy', 'Grade', 'Mutation.Count', 'Percent.aneuploidy', 'IDH.status'], axis = 1)\n",
    "    X = data_clean\n",
    "\n",
    "    preproc_scaler = load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "                                 file = 'scaler.joblib'\n",
    "                                )\n",
    "    # preprocess X_train, X_test and X_val\n",
    "    X_pred= preproc_scaler.transform(X)\n",
    "\n",
    "    return X_pred\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def save_scaler(scaler_to_save = None,\n",
    "               scaler_type = None,\n",
    "               path_to_save = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers\"\n",
    "              ):\n",
    "    \"\"\"\n",
    "    Persist trained model locally on the hard drive at f\"{path_to_save/scaler_type/f\"{timestamp}.joblib\"\n",
    "    \"\"\"\n",
    "\n",
    "    timestamp = time.strftime(\"%Y%m%d-%H%M%S\")\n",
    "\n",
    "    # Save scaler locally\n",
    "    scaler_path_file = os.path.join(f\"{path_to_save}/{scaler_type}_{timestamp}.joblib\")\n",
    "\n",
    "    dump(scaler_to_save, scaler_path_file)\n",
    "\n",
    "    print(f\"âœ… Scaler saved locally at {scaler_path_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9a1ed6ec-6253-4ff4-a734-c323bfe2290d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Identifier</th>\n",
       "      <th>Case</th>\n",
       "      <th>years_to_birth</th>\n",
       "      <th>gender</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>radiation_therapy</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Mutation.Count</th>\n",
       "      <th>...</th>\n",
       "      <th>p27_p</th>\n",
       "      <th>p27_pT157_p</th>\n",
       "      <th>p27_pT198_p</th>\n",
       "      <th>p38_pT180_Y182_p</th>\n",
       "      <th>p53_p</th>\n",
       "      <th>p62-LCK-ligand_p</th>\n",
       "      <th>p70S6K_p</th>\n",
       "      <th>p70S6K_pT389_p</th>\n",
       "      <th>p90RSK_p</th>\n",
       "      <th>p90RSK_pT359_S363_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>274</td>\n",
       "      <td>TCGA-QH-A65V</td>\n",
       "      <td>43</td>\n",
       "      <td>female</td>\n",
       "      <td>oligodendroglioma</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>no</td>\n",
       "      <td>G2</td>\n",
       "      <td>22</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023014</td>\n",
       "      <td>0.145349</td>\n",
       "      <td>0.113206</td>\n",
       "      <td>-0.066341</td>\n",
       "      <td>-0.199884</td>\n",
       "      <td>-0.334506</td>\n",
       "      <td>0.013173</td>\n",
       "      <td>0.598493</td>\n",
       "      <td>0.020249</td>\n",
       "      <td>-0.037049</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 186 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Identifier          Case  years_to_birth  gender  histological_type   race  \\\n",
       "0         274  TCGA-QH-A65V              43  female  oligodendroglioma  white   \n",
       "\n",
       "                ethnicity radiation_therapy Grade  Mutation.Count  ...  \\\n",
       "0  not hispanic or latino                no    G2              22  ...   \n",
       "\n",
       "      p27_p p27_pT157_p  p27_pT198_p  p38_pT180_Y182_p     p53_p  \\\n",
       "0  0.023014    0.145349     0.113206         -0.066341 -0.199884   \n",
       "\n",
       "   p62-LCK-ligand_p  p70S6K_p  p70S6K_pT389_p  p90RSK_p  p90RSK_pT359_S363_p  \n",
       "0         -0.334506  0.013173        0.598493  0.020249            -0.037049  \n",
       "\n",
       "[1 rows x 186 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data/Glioma-clinic-TCGA-proteins-all-cols-test-with-identifier-outcome0.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "20f25787",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Preprocess input proteins, age and gender... \n",
      "âœ… Data cleaned\n"
     ]
    }
   ],
   "source": [
    "df = df.drop(columns=\"Identifier\", axis = 1)\n",
    "# Preprocess\n",
    "X_pred = preprocess_input(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "e6b3b003-af21-43a3-b7c2-a54a7c4d8774",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Predict data\n",
    "model = load_model(path ='/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_models',\n",
    "                    file = 'sgd_model.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "f0fd20e4-53dd-4c7d-9e21-b4df20317710",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1., 0.]])"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(model.predict(X_pred)[0])\n",
    "model.predict_proba(X_pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdc6726a-1b55-456a-b8d9-a1ee1004c4e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "outcome_num = int(model.predict(X_pred)[0])\n",
    "if outcome_num == 0:\n",
    "    outcome = \"good cancer\"\n",
    "    probability = round(float(model.predict_proba(X_pred)[0][0]), 4)\n",
    "else:\n",
    "    outcome = \"bad cancer\"\n",
    "    probability = round(float(model.predict_proba(X_pred)[0][1]), 4)\n",
    "\n",
    "return {\n",
    "            \"Outcome\": outcome,\n",
    "            \"Probability\": probability\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3424412b-9e5b-46c9-afcb-6d8cbc07cb8c",
   "metadata": {},
   "source": [
    "### Output with several lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b8aba69f-9257-4f94-a3f6-0bd0d1d426de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Preprocess input proteins, age and gender... \n",
      "âœ… Data cleaned\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Preprocess\n",
    "X_pred = preprocess_input(df)\n",
    "\n",
    "# Predict data\n",
    "model = load_model(path ='/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_models',\n",
    "                    file = 'sgd_model.pkl')\n",
    "\n",
    "outcome = pd.DataFrame(model.predict(X_pred), columns=[\"Outcome\"], dtype = int)\n",
    "prob = pd.DataFrame(model.predict_proba(X_pred), columns=[\"Probability_0\", \"Probability_1\"], dtype = float)\n",
    "\n",
    "# Merge results into dataframe\n",
    "result = pd.merge(prob,outcome, left_index=True, right_index=True)\n",
    "result_dict = result.to_dict('series')\n",
    "\n",
    "prediction = {k: v.tolist() for k, v in result.iterrows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "019e5f52-88d6-4cf9-b2a8-527375fc5639",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0\n",
       "0  1.0\n",
       "1  0.0\n",
       "2  0.0"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(prediction)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f7163d1",
   "metadata": {},
   "source": [
    "# Create prediction endpoint for API"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ec0fde0-8122-47b7-8d31-4835da0b2615",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## User can upload the data using the upload button"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1969733b-dcab-4c71-90fb-e8b542e6be6b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8d36caa0-3866-40ae-81b7-9dab219a5093",
   "metadata": {},
   "source": [
    "## User has to input the path and the file name to upload file to predict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8435164a-a88a-4e6b-8d32-378cdfd2f5d9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Prediction endpoint for one line of prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f51efcec-9e02-4e93-b620-1cae583b7ee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_one(\n",
    "            path: str, #\"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "            file: str, # \"Glioma-clinic-TCGA-proteins-test-with-identifier.csv\"\n",
    "            ):\n",
    "    \"\"\"\n",
    "    Make a prediction for every row in your dataset.\n",
    "    Input needs to be a csv file with rows = samples and columns = proteins.\n",
    "    The first row has to contain the protein names or any other identifier that will serve as a header.\n",
    "    If your file has a sample identifier column, name this column \"Identifier\"\n",
    "    \"\"\"\n",
    "\n",
    "    df = pd.read_csv(f\"{path}/{file}\", header=0)\n",
    "\n",
    "    if 'Identifier' in df.columns:\n",
    "        X_pred = df.drop([\"Identifier\"], axis = 1)\n",
    "    else:\n",
    "        X_pred = df\n",
    "\n",
    "\n",
    "    # Preprocess\n",
    "    # Load scaler\n",
    "    scaler = load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "                         file = 'MinMax_20240306-102844.joblib'\n",
    "                        )\n",
    "\n",
    "    X_pred_proc = scaler.transform(X_pred)\n",
    "\n",
    "    # Predict data\n",
    "    model = load_model()\n",
    "\n",
    "    outcome_num = int(model.predict(X_pred_proc)[0])\n",
    "    if outcome_num == 0:\n",
    "        outcome = \"good cancer\"\n",
    "        probability = round(float(model.predict_proba(X_pred_proc)[0][0]), 10)\n",
    "    else:\n",
    "        outcome = \"bad cancer\"\n",
    "        probability = round(float(model.predict_proba(X_pred_proc)[0][1]), 10)\n",
    "\n",
    "    return {\n",
    "                \"Outcome\": outcome,\n",
    "                \"Probability\": probability\n",
    "            }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "dc107f9c-90bd-4e78-a2b7-f54a85be8399",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator MinMaxScaler from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Outcome': 'bad cancer', 'Probability': 1.0}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predict_one(path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\",\n",
    "       file = \"Glioma-clinic-TCGA-proteins-test-with-identifier-outcome1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b6bdf8ac-8f3f-450d-bb7c-dbdaec209ea3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "model = load_model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "93e4d539-31cb-4367-8bd9-efea6acf0f8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but SGDClassifier was fitted without feature names\n",
      "  warnings.warn(\n",
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but SGDClassifier was fitted without feature names\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "tmp1 = df[274:275].drop(cols_to_drop, axis = 1)\n",
    "outcome_num = int(model.predict(tmp1)[0])\n",
    "if outcome_num == 0:\n",
    "    outcome = \"good cancer\"\n",
    "    probability = round(float(model.predict_proba(tmp1)[0][0]), 4)\n",
    "else:\n",
    "    outcome = \"bad cancer\"\n",
    "    probability = round(float(model.predict_proba(tmp1)[0][1]), 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "3e161961-45fd-4e13-b500-895b35d30c5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Outcome': 'good cancer', 'Probability': 1.0}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{\n",
    "    \"Outcome\": outcome,\n",
    "    \"Probability\": probability\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9d7a4b2d-c49d-44a3-9ca9-1489f38930f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but SGDClassifier was fitted without feature names\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1.180455213004863e-08"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prob = float(model.predict_proba(tmp1)[0][1])\n",
    "prob"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42693c65-d7cb-4f1c-ad64-a718ba96265d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###  Prediction endpoint that should return a dataframe in case the user wants to do several predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d9bdb8bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# http://127.0.0.1:8000/predict?\n",
    "# path=/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\n",
    "# &file=Glioma-clinic-TCGA-proteins-test-with-identifier.csv\n",
    "@app.get(\"/predict\")\n",
    "def predict(\n",
    "            path: str, #\"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "            file: str, # \"Glioma-clinic-TCGA-proteins-test-with-identifier.csv\"\n",
    "            ):\n",
    "    \"\"\"\n",
    "    Make a prediction for every row in your dataset.\n",
    "    Input needs to be a csv file with rows = samples and columns = proteins.\n",
    "    The first row has to contain the protein names or any other identifier that will serve as a header.\n",
    "    If your file has a sample identifier column, name this column \"Identifier\"\n",
    "    \"\"\"\n",
    "\n",
    "    df = pd.read_csv(f\"{path}/{file}\", header=0)\n",
    "\n",
    "    if 'Identifier' in df.columns:\n",
    "        X_pred = df.drop([\"Identifier\"], axis = 1)\n",
    "    else:\n",
    "        X_pred = df\n",
    "\n",
    "\n",
    "    # Preprocess\n",
    "    # Load scaler\n",
    "    scaler = load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "                         file = 'MinMax_20240306-102844.joblib'\n",
    "                        )\n",
    "\n",
    "    X_pred_proc = scaler.transform(X_pred)\n",
    "\n",
    "    # Predict data\n",
    "    model = load_model()\n",
    "\n",
    "    outcome = pd.DataFrame(model.predict(X_pred_proc), columns=[\"Outcome\"], dtype = int)\n",
    "    prob = pd.DataFrame(model.predict_proba(X_pred_proc), columns=[\"Probability_0\", \"Probability_1\"], dtype = float)\n",
    "\n",
    "    # Merge results into dataframe\n",
    "    result = pd.merge(prob,outcome, left_index=True, right_index=True)\n",
    "    result_dict = result.to_dict('series')\n",
    "\n",
    "    return {k: v.tolist() for k, v in result.iterrows()}\n",
    "\n",
    "    #return {\"Outcome\": result[\"outcome\"][1]}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "bd5b1c41",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator MinMaxScaler from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "result = predict(path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\",\n",
    "       file = \"Glioma-clinic-TCGA-proteins-test-with-identifier.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72252809-a73e-44c5-8d1d-8cf79d50d111",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "##  Try to convert dataframe to be predicted to dictionary to pass into the api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "a7548400-f600-4ce6-90dc-2a602aeaae82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data to predict\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-test-with-identifier.csv\"\n",
    "\n",
    "df1 = pd.read_csv(f\"{path}/{file}\", header=0)\n",
    "\n",
    "if 'Identifier' in df1.columns:\n",
    "    X_pred = df1.drop([\"Identifier\"], axis = 1)\n",
    "else:\n",
    "    X_pred = df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d31b3843-b278-422d-b950-089a31fa06c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>14-3-3_beta_p</th>\n",
       "      <th>14-3-3_epsilon_p</th>\n",
       "      <th>14-3-3_zeta_p</th>\n",
       "      <th>4E-BP1_p</th>\n",
       "      <th>4E-BP1_pS65_p</th>\n",
       "      <th>4E-BP1_pT37_T46_p</th>\n",
       "      <th>4E-BP1_pT70_p</th>\n",
       "      <th>53BP1_p</th>\n",
       "      <th>A-Raf_pS299_p</th>\n",
       "      <th>ACC1_p</th>\n",
       "      <th>...</th>\n",
       "      <th>p27_p</th>\n",
       "      <th>p27_pT157_p</th>\n",
       "      <th>p27_pT198_p</th>\n",
       "      <th>p38_pT180_Y182_p</th>\n",
       "      <th>p53_p</th>\n",
       "      <th>p62-LCK-ligand_p</th>\n",
       "      <th>p70S6K_p</th>\n",
       "      <th>p70S6K_pT389_p</th>\n",
       "      <th>p90RSK_p</th>\n",
       "      <th>p90RSK_pT359_S363_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.046266</td>\n",
       "      <td>0.010482</td>\n",
       "      <td>0.005867</td>\n",
       "      <td>-0.128982</td>\n",
       "      <td>-0.097890</td>\n",
       "      <td>-0.286694</td>\n",
       "      <td>-0.095360</td>\n",
       "      <td>-0.315572</td>\n",
       "      <td>-0.079862</td>\n",
       "      <td>-0.190765</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282454</td>\n",
       "      <td>0.215619</td>\n",
       "      <td>0.049541</td>\n",
       "      <td>0.394208</td>\n",
       "      <td>-0.173627</td>\n",
       "      <td>-0.088369</td>\n",
       "      <td>-0.160519</td>\n",
       "      <td>0.494404</td>\n",
       "      <td>0.412308</td>\n",
       "      <td>-0.081711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.062830</td>\n",
       "      <td>0.341067</td>\n",
       "      <td>0.509924</td>\n",
       "      <td>-0.097006</td>\n",
       "      <td>0.050285</td>\n",
       "      <td>-0.555042</td>\n",
       "      <td>-0.108214</td>\n",
       "      <td>-0.047927</td>\n",
       "      <td>-0.029663</td>\n",
       "      <td>0.203558</td>\n",
       "      <td>...</td>\n",
       "      <td>0.233966</td>\n",
       "      <td>0.165461</td>\n",
       "      <td>0.225253</td>\n",
       "      <td>-0.065668</td>\n",
       "      <td>-0.268971</td>\n",
       "      <td>-0.059341</td>\n",
       "      <td>-0.237159</td>\n",
       "      <td>-0.004258</td>\n",
       "      <td>0.383249</td>\n",
       "      <td>-0.383407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.005718</td>\n",
       "      <td>-0.049226</td>\n",
       "      <td>-0.176334</td>\n",
       "      <td>0.110042</td>\n",
       "      <td>0.084026</td>\n",
       "      <td>0.205933</td>\n",
       "      <td>-0.054664</td>\n",
       "      <td>0.562995</td>\n",
       "      <td>0.063561</td>\n",
       "      <td>0.578327</td>\n",
       "      <td>...</td>\n",
       "      <td>0.361402</td>\n",
       "      <td>-0.143755</td>\n",
       "      <td>-0.065628</td>\n",
       "      <td>0.216761</td>\n",
       "      <td>0.033749</td>\n",
       "      <td>0.105418</td>\n",
       "      <td>-0.248534</td>\n",
       "      <td>-0.251022</td>\n",
       "      <td>0.017531</td>\n",
       "      <td>-0.108484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.070416</td>\n",
       "      <td>0.068269</td>\n",
       "      <td>0.534429</td>\n",
       "      <td>-0.403894</td>\n",
       "      <td>0.106003</td>\n",
       "      <td>-0.169416</td>\n",
       "      <td>-0.234652</td>\n",
       "      <td>-0.223324</td>\n",
       "      <td>0.019363</td>\n",
       "      <td>0.397784</td>\n",
       "      <td>...</td>\n",
       "      <td>0.113838</td>\n",
       "      <td>-0.081752</td>\n",
       "      <td>-0.028470</td>\n",
       "      <td>0.059929</td>\n",
       "      <td>-0.093907</td>\n",
       "      <td>0.007829</td>\n",
       "      <td>-0.040814</td>\n",
       "      <td>0.726636</td>\n",
       "      <td>0.156300</td>\n",
       "      <td>-0.195128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.063005</td>\n",
       "      <td>0.047834</td>\n",
       "      <td>0.414731</td>\n",
       "      <td>-0.374914</td>\n",
       "      <td>0.616939</td>\n",
       "      <td>0.810131</td>\n",
       "      <td>-0.173296</td>\n",
       "      <td>-0.381965</td>\n",
       "      <td>0.101445</td>\n",
       "      <td>0.273052</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.273558</td>\n",
       "      <td>-0.042558</td>\n",
       "      <td>-0.142239</td>\n",
       "      <td>-0.141052</td>\n",
       "      <td>-0.224926</td>\n",
       "      <td>-0.073704</td>\n",
       "      <td>-0.076397</td>\n",
       "      <td>1.391063</td>\n",
       "      <td>0.137433</td>\n",
       "      <td>0.104366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>0.163980</td>\n",
       "      <td>0.029452</td>\n",
       "      <td>0.649739</td>\n",
       "      <td>-0.445981</td>\n",
       "      <td>1.062340</td>\n",
       "      <td>1.745698</td>\n",
       "      <td>0.025130</td>\n",
       "      <td>-0.152001</td>\n",
       "      <td>0.239102</td>\n",
       "      <td>0.431975</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.104785</td>\n",
       "      <td>0.252492</td>\n",
       "      <td>-0.155413</td>\n",
       "      <td>-0.746732</td>\n",
       "      <td>0.029633</td>\n",
       "      <td>0.219986</td>\n",
       "      <td>0.093239</td>\n",
       "      <td>2.279390</td>\n",
       "      <td>0.222581</td>\n",
       "      <td>0.341038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>-0.029360</td>\n",
       "      <td>-0.020617</td>\n",
       "      <td>-0.058948</td>\n",
       "      <td>0.173444</td>\n",
       "      <td>0.115165</td>\n",
       "      <td>0.235074</td>\n",
       "      <td>0.065090</td>\n",
       "      <td>-0.202499</td>\n",
       "      <td>0.023789</td>\n",
       "      <td>0.080197</td>\n",
       "      <td>...</td>\n",
       "      <td>0.422843</td>\n",
       "      <td>-0.034226</td>\n",
       "      <td>0.087660</td>\n",
       "      <td>-0.013897</td>\n",
       "      <td>0.336788</td>\n",
       "      <td>0.012812</td>\n",
       "      <td>0.066353</td>\n",
       "      <td>-0.132282</td>\n",
       "      <td>-0.191353</td>\n",
       "      <td>0.017849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>-0.082008</td>\n",
       "      <td>0.253038</td>\n",
       "      <td>0.390207</td>\n",
       "      <td>-0.166794</td>\n",
       "      <td>-0.119515</td>\n",
       "      <td>-0.566232</td>\n",
       "      <td>-0.140786</td>\n",
       "      <td>-0.212030</td>\n",
       "      <td>0.160434</td>\n",
       "      <td>0.107996</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.079885</td>\n",
       "      <td>0.121889</td>\n",
       "      <td>0.062165</td>\n",
       "      <td>-0.484047</td>\n",
       "      <td>0.274336</td>\n",
       "      <td>0.177687</td>\n",
       "      <td>0.064581</td>\n",
       "      <td>1.263077</td>\n",
       "      <td>0.215282</td>\n",
       "      <td>0.108698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>0.058965</td>\n",
       "      <td>-0.050453</td>\n",
       "      <td>0.028261</td>\n",
       "      <td>-0.050225</td>\n",
       "      <td>-0.221785</td>\n",
       "      <td>-0.788873</td>\n",
       "      <td>-0.199018</td>\n",
       "      <td>-0.101192</td>\n",
       "      <td>0.108723</td>\n",
       "      <td>-0.416462</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.063589</td>\n",
       "      <td>-0.046370</td>\n",
       "      <td>0.004108</td>\n",
       "      <td>-0.000405</td>\n",
       "      <td>0.283721</td>\n",
       "      <td>0.420645</td>\n",
       "      <td>-0.148972</td>\n",
       "      <td>-0.176567</td>\n",
       "      <td>-0.095915</td>\n",
       "      <td>0.066603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>-0.089893</td>\n",
       "      <td>0.184818</td>\n",
       "      <td>-0.379278</td>\n",
       "      <td>-0.113470</td>\n",
       "      <td>-0.381345</td>\n",
       "      <td>0.469686</td>\n",
       "      <td>-0.360194</td>\n",
       "      <td>-0.061591</td>\n",
       "      <td>-0.033238</td>\n",
       "      <td>-0.530723</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.295703</td>\n",
       "      <td>-0.126190</td>\n",
       "      <td>0.002579</td>\n",
       "      <td>0.682083</td>\n",
       "      <td>0.248918</td>\n",
       "      <td>-0.268641</td>\n",
       "      <td>-0.000092</td>\n",
       "      <td>-1.126171</td>\n",
       "      <td>-0.029505</td>\n",
       "      <td>0.079091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows Ã— 174 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    14-3-3_beta_p  14-3-3_epsilon_p  14-3-3_zeta_p  4E-BP1_p  4E-BP1_pS65_p  \\\n",
       "0       -0.046266          0.010482       0.005867 -0.128982      -0.097890   \n",
       "1        0.062830          0.341067       0.509924 -0.097006       0.050285   \n",
       "2        0.005718         -0.049226      -0.176334  0.110042       0.084026   \n",
       "3        0.070416          0.068269       0.534429 -0.403894       0.106003   \n",
       "4       -0.063005          0.047834       0.414731 -0.374914       0.616939   \n",
       "..            ...               ...            ...       ...            ...   \n",
       "57       0.163980          0.029452       0.649739 -0.445981       1.062340   \n",
       "58      -0.029360         -0.020617      -0.058948  0.173444       0.115165   \n",
       "59      -0.082008          0.253038       0.390207 -0.166794      -0.119515   \n",
       "60       0.058965         -0.050453       0.028261 -0.050225      -0.221785   \n",
       "61      -0.089893          0.184818      -0.379278 -0.113470      -0.381345   \n",
       "\n",
       "    4E-BP1_pT37_T46_p  4E-BP1_pT70_p   53BP1_p  A-Raf_pS299_p    ACC1_p  ...  \\\n",
       "0           -0.286694      -0.095360 -0.315572      -0.079862 -0.190765  ...   \n",
       "1           -0.555042      -0.108214 -0.047927      -0.029663  0.203558  ...   \n",
       "2            0.205933      -0.054664  0.562995       0.063561  0.578327  ...   \n",
       "3           -0.169416      -0.234652 -0.223324       0.019363  0.397784  ...   \n",
       "4            0.810131      -0.173296 -0.381965       0.101445  0.273052  ...   \n",
       "..                ...            ...       ...            ...       ...  ...   \n",
       "57           1.745698       0.025130 -0.152001       0.239102  0.431975  ...   \n",
       "58           0.235074       0.065090 -0.202499       0.023789  0.080197  ...   \n",
       "59          -0.566232      -0.140786 -0.212030       0.160434  0.107996  ...   \n",
       "60          -0.788873      -0.199018 -0.101192       0.108723 -0.416462  ...   \n",
       "61           0.469686      -0.360194 -0.061591      -0.033238 -0.530723  ...   \n",
       "\n",
       "       p27_p  p27_pT157_p  p27_pT198_p  p38_pT180_Y182_p     p53_p  \\\n",
       "0   0.282454     0.215619     0.049541          0.394208 -0.173627   \n",
       "1   0.233966     0.165461     0.225253         -0.065668 -0.268971   \n",
       "2   0.361402    -0.143755    -0.065628          0.216761  0.033749   \n",
       "3   0.113838    -0.081752    -0.028470          0.059929 -0.093907   \n",
       "4  -0.273558    -0.042558    -0.142239         -0.141052 -0.224926   \n",
       "..       ...          ...          ...               ...       ...   \n",
       "57 -0.104785     0.252492    -0.155413         -0.746732  0.029633   \n",
       "58  0.422843    -0.034226     0.087660         -0.013897  0.336788   \n",
       "59 -0.079885     0.121889     0.062165         -0.484047  0.274336   \n",
       "60 -0.063589    -0.046370     0.004108         -0.000405  0.283721   \n",
       "61 -0.295703    -0.126190     0.002579          0.682083  0.248918   \n",
       "\n",
       "    p62-LCK-ligand_p  p70S6K_p  p70S6K_pT389_p  p90RSK_p  p90RSK_pT359_S363_p  \n",
       "0          -0.088369 -0.160519        0.494404  0.412308            -0.081711  \n",
       "1          -0.059341 -0.237159       -0.004258  0.383249            -0.383407  \n",
       "2           0.105418 -0.248534       -0.251022  0.017531            -0.108484  \n",
       "3           0.007829 -0.040814        0.726636  0.156300            -0.195128  \n",
       "4          -0.073704 -0.076397        1.391063  0.137433             0.104366  \n",
       "..               ...       ...             ...       ...                  ...  \n",
       "57          0.219986  0.093239        2.279390  0.222581             0.341038  \n",
       "58          0.012812  0.066353       -0.132282 -0.191353             0.017849  \n",
       "59          0.177687  0.064581        1.263077  0.215282             0.108698  \n",
       "60          0.420645 -0.148972       -0.176567 -0.095915             0.066603  \n",
       "61         -0.268641 -0.000092       -1.126171 -0.029505             0.079091  \n",
       "\n",
       "[62 rows x 174 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "ec4976b0-759d-4607-90cb-31dd18cc185f",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_pred_list = {k: v.tolist() for k, v in X_pred.iterrows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "c7dba2ba-e98e-4df9-b3ac-d8437fb541fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>52</th>\n",
       "      <th>53</th>\n",
       "      <th>54</th>\n",
       "      <th>55</th>\n",
       "      <th>56</th>\n",
       "      <th>57</th>\n",
       "      <th>58</th>\n",
       "      <th>59</th>\n",
       "      <th>60</th>\n",
       "      <th>61</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.046266</td>\n",
       "      <td>0.062830</td>\n",
       "      <td>0.005718</td>\n",
       "      <td>0.070416</td>\n",
       "      <td>-0.063005</td>\n",
       "      <td>0.090642</td>\n",
       "      <td>-0.093869</td>\n",
       "      <td>0.117048</td>\n",
       "      <td>-0.036813</td>\n",
       "      <td>-0.188983</td>\n",
       "      <td>...</td>\n",
       "      <td>0.129091</td>\n",
       "      <td>0.188977</td>\n",
       "      <td>-0.092673</td>\n",
       "      <td>-0.070395</td>\n",
       "      <td>0.088773</td>\n",
       "      <td>0.163980</td>\n",
       "      <td>-0.029360</td>\n",
       "      <td>-0.082008</td>\n",
       "      <td>0.058965</td>\n",
       "      <td>-0.089893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.010482</td>\n",
       "      <td>0.341067</td>\n",
       "      <td>-0.049226</td>\n",
       "      <td>0.068269</td>\n",
       "      <td>0.047834</td>\n",
       "      <td>0.127904</td>\n",
       "      <td>-0.252845</td>\n",
       "      <td>0.241715</td>\n",
       "      <td>-0.047815</td>\n",
       "      <td>-0.311200</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004169</td>\n",
       "      <td>-0.058367</td>\n",
       "      <td>-0.135226</td>\n",
       "      <td>-0.034295</td>\n",
       "      <td>0.054260</td>\n",
       "      <td>0.029452</td>\n",
       "      <td>-0.020617</td>\n",
       "      <td>0.253038</td>\n",
       "      <td>-0.050453</td>\n",
       "      <td>0.184818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.005867</td>\n",
       "      <td>0.509924</td>\n",
       "      <td>-0.176334</td>\n",
       "      <td>0.534429</td>\n",
       "      <td>0.414731</td>\n",
       "      <td>0.339170</td>\n",
       "      <td>0.410162</td>\n",
       "      <td>0.206968</td>\n",
       "      <td>-0.122072</td>\n",
       "      <td>-0.225733</td>\n",
       "      <td>...</td>\n",
       "      <td>0.098866</td>\n",
       "      <td>-0.111915</td>\n",
       "      <td>-0.120118</td>\n",
       "      <td>0.054421</td>\n",
       "      <td>-0.140119</td>\n",
       "      <td>0.649739</td>\n",
       "      <td>-0.058948</td>\n",
       "      <td>0.390207</td>\n",
       "      <td>0.028261</td>\n",
       "      <td>-0.379278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.128982</td>\n",
       "      <td>-0.097006</td>\n",
       "      <td>0.110042</td>\n",
       "      <td>-0.403894</td>\n",
       "      <td>-0.374914</td>\n",
       "      <td>-0.150820</td>\n",
       "      <td>-0.327964</td>\n",
       "      <td>0.007823</td>\n",
       "      <td>-0.050649</td>\n",
       "      <td>0.342750</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.162787</td>\n",
       "      <td>-0.212003</td>\n",
       "      <td>0.220188</td>\n",
       "      <td>-0.266592</td>\n",
       "      <td>0.037902</td>\n",
       "      <td>-0.445981</td>\n",
       "      <td>0.173444</td>\n",
       "      <td>-0.166794</td>\n",
       "      <td>-0.050225</td>\n",
       "      <td>-0.113470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.097890</td>\n",
       "      <td>0.050285</td>\n",
       "      <td>0.084026</td>\n",
       "      <td>0.106003</td>\n",
       "      <td>0.616939</td>\n",
       "      <td>0.365706</td>\n",
       "      <td>0.197215</td>\n",
       "      <td>-0.062358</td>\n",
       "      <td>-0.220442</td>\n",
       "      <td>0.241693</td>\n",
       "      <td>...</td>\n",
       "      <td>0.229510</td>\n",
       "      <td>-0.164837</td>\n",
       "      <td>0.106521</td>\n",
       "      <td>-0.158487</td>\n",
       "      <td>-0.313040</td>\n",
       "      <td>1.062340</td>\n",
       "      <td>0.115165</td>\n",
       "      <td>-0.119515</td>\n",
       "      <td>-0.221785</td>\n",
       "      <td>-0.381345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>-0.088369</td>\n",
       "      <td>-0.059341</td>\n",
       "      <td>0.105418</td>\n",
       "      <td>0.007829</td>\n",
       "      <td>-0.073704</td>\n",
       "      <td>0.004746</td>\n",
       "      <td>0.340390</td>\n",
       "      <td>0.035733</td>\n",
       "      <td>0.250815</td>\n",
       "      <td>-0.466182</td>\n",
       "      <td>...</td>\n",
       "      <td>0.087680</td>\n",
       "      <td>-0.095666</td>\n",
       "      <td>-0.474115</td>\n",
       "      <td>0.043443</td>\n",
       "      <td>0.254037</td>\n",
       "      <td>0.219986</td>\n",
       "      <td>0.012812</td>\n",
       "      <td>0.177687</td>\n",
       "      <td>0.420645</td>\n",
       "      <td>-0.268641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>-0.160519</td>\n",
       "      <td>-0.237159</td>\n",
       "      <td>-0.248534</td>\n",
       "      <td>-0.040814</td>\n",
       "      <td>-0.076397</td>\n",
       "      <td>-0.039065</td>\n",
       "      <td>0.020190</td>\n",
       "      <td>-0.045691</td>\n",
       "      <td>-0.048886</td>\n",
       "      <td>0.247032</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.100576</td>\n",
       "      <td>-0.122288</td>\n",
       "      <td>0.116780</td>\n",
       "      <td>-0.043661</td>\n",
       "      <td>-0.230895</td>\n",
       "      <td>0.093239</td>\n",
       "      <td>0.066353</td>\n",
       "      <td>0.064581</td>\n",
       "      <td>-0.148972</td>\n",
       "      <td>-0.000092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>0.494404</td>\n",
       "      <td>-0.004258</td>\n",
       "      <td>-0.251022</td>\n",
       "      <td>0.726636</td>\n",
       "      <td>1.391063</td>\n",
       "      <td>2.322902</td>\n",
       "      <td>0.183455</td>\n",
       "      <td>0.030409</td>\n",
       "      <td>0.389122</td>\n",
       "      <td>-0.111868</td>\n",
       "      <td>...</td>\n",
       "      <td>1.727192</td>\n",
       "      <td>-0.123695</td>\n",
       "      <td>-0.503557</td>\n",
       "      <td>0.325470</td>\n",
       "      <td>-0.116572</td>\n",
       "      <td>2.279390</td>\n",
       "      <td>-0.132282</td>\n",
       "      <td>1.263077</td>\n",
       "      <td>-0.176567</td>\n",
       "      <td>-1.126171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>0.412308</td>\n",
       "      <td>0.383249</td>\n",
       "      <td>0.017531</td>\n",
       "      <td>0.156300</td>\n",
       "      <td>0.137433</td>\n",
       "      <td>0.151344</td>\n",
       "      <td>0.125674</td>\n",
       "      <td>0.030682</td>\n",
       "      <td>-0.001712</td>\n",
       "      <td>-0.047198</td>\n",
       "      <td>...</td>\n",
       "      <td>0.156863</td>\n",
       "      <td>-0.164060</td>\n",
       "      <td>-0.039171</td>\n",
       "      <td>0.111535</td>\n",
       "      <td>-0.196836</td>\n",
       "      <td>0.222581</td>\n",
       "      <td>-0.191353</td>\n",
       "      <td>0.215282</td>\n",
       "      <td>-0.095915</td>\n",
       "      <td>-0.029505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>-0.081711</td>\n",
       "      <td>-0.383407</td>\n",
       "      <td>-0.108484</td>\n",
       "      <td>-0.195128</td>\n",
       "      <td>0.104366</td>\n",
       "      <td>-0.175998</td>\n",
       "      <td>0.062502</td>\n",
       "      <td>-0.035506</td>\n",
       "      <td>0.054310</td>\n",
       "      <td>0.102747</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.213446</td>\n",
       "      <td>-0.121061</td>\n",
       "      <td>-0.117333</td>\n",
       "      <td>-0.221823</td>\n",
       "      <td>0.151965</td>\n",
       "      <td>0.341038</td>\n",
       "      <td>0.017849</td>\n",
       "      <td>0.108698</td>\n",
       "      <td>0.066603</td>\n",
       "      <td>0.079091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>174 rows Ã— 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2         3         4         5         6   \\\n",
       "0   -0.046266  0.062830  0.005718  0.070416 -0.063005  0.090642 -0.093869   \n",
       "1    0.010482  0.341067 -0.049226  0.068269  0.047834  0.127904 -0.252845   \n",
       "2    0.005867  0.509924 -0.176334  0.534429  0.414731  0.339170  0.410162   \n",
       "3   -0.128982 -0.097006  0.110042 -0.403894 -0.374914 -0.150820 -0.327964   \n",
       "4   -0.097890  0.050285  0.084026  0.106003  0.616939  0.365706  0.197215   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "169 -0.088369 -0.059341  0.105418  0.007829 -0.073704  0.004746  0.340390   \n",
       "170 -0.160519 -0.237159 -0.248534 -0.040814 -0.076397 -0.039065  0.020190   \n",
       "171  0.494404 -0.004258 -0.251022  0.726636  1.391063  2.322902  0.183455   \n",
       "172  0.412308  0.383249  0.017531  0.156300  0.137433  0.151344  0.125674   \n",
       "173 -0.081711 -0.383407 -0.108484 -0.195128  0.104366 -0.175998  0.062502   \n",
       "\n",
       "           7         8         9   ...        52        53        54  \\\n",
       "0    0.117048 -0.036813 -0.188983  ...  0.129091  0.188977 -0.092673   \n",
       "1    0.241715 -0.047815 -0.311200  ...  0.004169 -0.058367 -0.135226   \n",
       "2    0.206968 -0.122072 -0.225733  ...  0.098866 -0.111915 -0.120118   \n",
       "3    0.007823 -0.050649  0.342750  ... -0.162787 -0.212003  0.220188   \n",
       "4   -0.062358 -0.220442  0.241693  ...  0.229510 -0.164837  0.106521   \n",
       "..        ...       ...       ...  ...       ...       ...       ...   \n",
       "169  0.035733  0.250815 -0.466182  ...  0.087680 -0.095666 -0.474115   \n",
       "170 -0.045691 -0.048886  0.247032  ... -0.100576 -0.122288  0.116780   \n",
       "171  0.030409  0.389122 -0.111868  ...  1.727192 -0.123695 -0.503557   \n",
       "172  0.030682 -0.001712 -0.047198  ...  0.156863 -0.164060 -0.039171   \n",
       "173 -0.035506  0.054310  0.102747  ... -0.213446 -0.121061 -0.117333   \n",
       "\n",
       "           55        56        57        58        59        60        61  \n",
       "0   -0.070395  0.088773  0.163980 -0.029360 -0.082008  0.058965 -0.089893  \n",
       "1   -0.034295  0.054260  0.029452 -0.020617  0.253038 -0.050453  0.184818  \n",
       "2    0.054421 -0.140119  0.649739 -0.058948  0.390207  0.028261 -0.379278  \n",
       "3   -0.266592  0.037902 -0.445981  0.173444 -0.166794 -0.050225 -0.113470  \n",
       "4   -0.158487 -0.313040  1.062340  0.115165 -0.119515 -0.221785 -0.381345  \n",
       "..        ...       ...       ...       ...       ...       ...       ...  \n",
       "169  0.043443  0.254037  0.219986  0.012812  0.177687  0.420645 -0.268641  \n",
       "170 -0.043661 -0.230895  0.093239  0.066353  0.064581 -0.148972 -0.000092  \n",
       "171  0.325470 -0.116572  2.279390 -0.132282  1.263077 -0.176567 -1.126171  \n",
       "172  0.111535 -0.196836  0.222581 -0.191353  0.215282 -0.095915 -0.029505  \n",
       "173 -0.221823  0.151965  0.341038  0.017849  0.108698  0.066603  0.079091  \n",
       "\n",
       "[174 rows x 62 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(X_pred_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff9fd9a4-0158-4cda-9cf7-bec00bf60bf0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "263b888d-69ba-4d9e-8413-a7d64ab70e27",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "40b608ef-23dc-42b8-9769-16178669ec1c",
   "metadata": {},
   "source": [
    "# Call API from website"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1eafce4-3e0a-453f-8448-230eb0126ca9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "##  Using path and file name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "30d6a8d2-1466-4e48-94e7-ce610831bb6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data'\n",
    "file = 'Glioma-clinic-TCGA-proteins-test-with-identifier-outcome1.csv'\n",
    "\n",
    "\n",
    "params = {'path': path,\n",
    "          'file': file}\n",
    "brainproteomics_api_url = 'http://127.0.0.1:8000/predict_one'\n",
    "response = requests.get(brainproteomics_api_url, params = params)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c36560f9-6c54-4e65-9257-89202f9a7efd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "578a9f0d-54a7-4811-94c0-3ada00cd11d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Outcome': 'bad cancer', 'Probability': 1.0}"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.json()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d48bb988-7aba-4a05-92a4-6ee2a92abf0c",
   "metadata": {},
   "source": [
    "## Using uploaded file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fd2c35f6-4108-4947-b5b3-c05f47830e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data to predict\n",
    "path = \"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data\"\n",
    "file = \"Glioma-clinic-TCGA-proteins-test-with-identifier.csv\"\n",
    "\n",
    "df = pd.read_csv(f\"{path}/{file}\", header=0)\n",
    "\n",
    "X_pred = df.iloc[:,1:]\n",
    "\n",
    "dict = {k: v.tolist() for k, v in X_pred.iterrows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "0b173405-286b-46c5-8b05-0d0ab7703177",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_one_upload(\n",
    "            dict # \"Glioma-clinic-TCGA-proteins-test-with-identifier-outcome0.csv\"\n",
    "            ):\n",
    "    \"\"\"\n",
    "    Make a prediction for every row in your dataset.\n",
    "    Input needs to be a csv file with rows = samples and columns = proteins.\n",
    "    The first row has to contain the protein names or any other identifier that will serve as a header.\n",
    "    If your file has a sample identifier column, name this column \"Identifier\"\n",
    "    \"\"\"\n",
    "\n",
    "    X_pred = pd.DataFrame(dict).transpose()\n",
    "\n",
    "\n",
    "    # Preprocess\n",
    "    # Load scaler\n",
    "    scaler = load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "                         file = 'MinMax_20240306-102844.joblib'\n",
    "                        )\n",
    "\n",
    "    X_pred_proc = scaler.transform(X_pred)\n",
    "\n",
    "    # Predict data\n",
    "    model = load_model()\n",
    "\n",
    "    outcome_num = int(model.predict(X_pred_proc)[0])\n",
    "    if outcome_num == 0:\n",
    "        outcome = \"good cancer\"\n",
    "        probability = round(float(model.predict_proba(X_pred_proc)[0][0]), 4)\n",
    "    else:\n",
    "        outcome = \"bad cancer\"\n",
    "        probability = round(float(model.predict_proba(X_pred_proc)[0][1]), 4)\n",
    "\n",
    "    return {\n",
    "                \"Outcome\": outcome,\n",
    "                \"Probability\": probability\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "f7044d3b-8204-4207-a54a-85449f2f7db8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator MinMaxScaler from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MinMaxScaler was fitted with feature names\n",
      "  warnings.warn(\n",
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Outcome': 'bad cancer', 'Probability': 1.0}"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predict_one_upload(dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "29b9bab2-603a-4dac-8da7-234072bb0d9c",
   "metadata": {},
   "outputs": [
    {
     "ename": "ChunkedEncodingError",
     "evalue": "(\"Connection broken: ConnectionResetError(104, 'Connection reset by peer')\", ConnectionResetError(104, 'Connection reset by peer'))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mConnectionResetError\u001b[0m                      Traceback (most recent call last)",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/urllib3/response.py:737\u001b[0m, in \u001b[0;36mHTTPResponse._error_catcher\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    736\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 737\u001b[0m     \u001b[38;5;28;01myield\u001b[39;00m\n\u001b[1;32m    739\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SocketTimeout \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    740\u001b[0m     \u001b[38;5;66;03m# FIXME: Ideally we'd like to include the url in the ReadTimeoutError but\u001b[39;00m\n\u001b[1;32m    741\u001b[0m     \u001b[38;5;66;03m# there is yet no clean way to get at it from this context.\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/urllib3/response.py:862\u001b[0m, in \u001b[0;36mHTTPResponse._raw_read\u001b[0;34m(self, amt, read1)\u001b[0m\n\u001b[1;32m    861\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_error_catcher():\n\u001b[0;32m--> 862\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mread1\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mread1\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m fp_closed \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    863\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m amt \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m data:\n\u001b[1;32m    864\u001b[0m         \u001b[38;5;66;03m# Platform-specific: Buggy versions of Python.\u001b[39;00m\n\u001b[1;32m    865\u001b[0m         \u001b[38;5;66;03m# Close the connection when no data is returned\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    870\u001b[0m         \u001b[38;5;66;03m# not properly close the connection in all cases. There is\u001b[39;00m\n\u001b[1;32m    871\u001b[0m         \u001b[38;5;66;03m# no harm in redundantly calling close.\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/urllib3/response.py:845\u001b[0m, in \u001b[0;36mHTTPResponse._fp_read\u001b[0;34m(self, amt, read1)\u001b[0m\n\u001b[1;32m    843\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    844\u001b[0m     \u001b[38;5;66;03m# StringIO doesn't like amt=None\u001b[39;00m\n\u001b[0;32m--> 845\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp\u001b[38;5;241m.\u001b[39mread()\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/http/client.py:465\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    464\u001b[0m     amt \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength\n\u001b[0;32m--> 465\u001b[0m s \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    466\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m s \u001b[38;5;129;01mand\u001b[39;00m amt:\n\u001b[1;32m    467\u001b[0m     \u001b[38;5;66;03m# Ideally, we would raise IncompleteRead if the content-length\u001b[39;00m\n\u001b[1;32m    468\u001b[0m     \u001b[38;5;66;03m# wasn't satisfied, but it might break compatibility.\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    704\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    706\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n",
      "\u001b[0;31mConnectionResetError\u001b[0m: [Errno 104] Connection reset by peer",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mProtocolError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/requests/models.py:816\u001b[0m, in \u001b[0;36mResponse.iter_content.<locals>.generate\u001b[0;34m()\u001b[0m\n\u001b[1;32m    815\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 816\u001b[0m     \u001b[38;5;28;01myield from\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mstream(chunk_size, decode_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    817\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ProtocolError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/urllib3/response.py:1043\u001b[0m, in \u001b[0;36mHTTPResponse.stream\u001b[0;34m(self, amt, decode_content)\u001b[0m\n\u001b[1;32m   1042\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_fp_closed(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_decoded_buffer) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m-> 1043\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mamt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1045\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m data:\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/urllib3/response.py:935\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[0;34m(self, amt, decode_content, cache_content)\u001b[0m\n\u001b[1;32m    933\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_decoded_buffer\u001b[38;5;241m.\u001b[39mget(amt)\n\u001b[0;32m--> 935\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raw_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    937\u001b[0m flush_decoder \u001b[38;5;241m=\u001b[39m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mor\u001b[39;00m (amt \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m data)\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/urllib3/response.py:861\u001b[0m, in \u001b[0;36mHTTPResponse._raw_read\u001b[0;34m(self, amt, read1)\u001b[0m\n\u001b[1;32m    859\u001b[0m fp_closed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclosed\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m--> 861\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_error_catcher():\n\u001b[1;32m    862\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp_read(amt, read1\u001b[38;5;241m=\u001b[39mread1) \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m fp_closed \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/contextlib.py:153\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__exit__\u001b[0;34m(self, typ, value, traceback)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 153\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgen\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mthrow\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtraceback\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    154\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    155\u001b[0m     \u001b[38;5;66;03m# Suppress StopIteration *unless* it's the same exception that\u001b[39;00m\n\u001b[1;32m    156\u001b[0m     \u001b[38;5;66;03m# was passed to throw().  This prevents a StopIteration\u001b[39;00m\n\u001b[1;32m    157\u001b[0m     \u001b[38;5;66;03m# raised inside the \"with\" statement from being suppressed.\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/urllib3/response.py:764\u001b[0m, in \u001b[0;36mHTTPResponse._error_catcher\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    763\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (HTTPException, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 764\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ProtocolError(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnection broken: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, e) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n\u001b[1;32m    766\u001b[0m \u001b[38;5;66;03m# If no exception is thrown, we should avoid cleaning up\u001b[39;00m\n\u001b[1;32m    767\u001b[0m \u001b[38;5;66;03m# unnecessarily.\u001b[39;00m\n",
      "\u001b[0;31mProtocolError\u001b[0m: (\"Connection broken: ConnectionResetError(104, 'Connection reset by peer')\", ConnectionResetError(104, 'Connection reset by peer'))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mChunkedEncodingError\u001b[0m                      Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 9\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28mdict\u001b[39m \u001b[38;5;241m=\u001b[39m {k: v\u001b[38;5;241m.\u001b[39mtolist() \u001b[38;5;28;01mfor\u001b[39;00m k, v \u001b[38;5;129;01min\u001b[39;00m X_pred\u001b[38;5;241m.\u001b[39miterrows()}\n\u001b[1;32m      8\u001b[0m brainproteomics_api_url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttp://127.0.0.1:8000/predict_one_upload\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 9\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbrainproteomics_api_url\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mdict\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     11\u001b[0m prediction \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mjson()\n\u001b[1;32m     13\u001b[0m st\u001b[38;5;241m.\u001b[39mwrite(prediction)\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/requests/sessions.py:747\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    744\u001b[0m         \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m    746\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m stream:\n\u001b[0;32m--> 747\u001b[0m     \u001b[43mr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcontent\u001b[49m\n\u001b[1;32m    749\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m r\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/requests/models.py:899\u001b[0m, in \u001b[0;36mResponse.content\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    897\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_content \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    898\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 899\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_content \u001b[38;5;241m=\u001b[39m \u001b[38;5;124;43mb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43miter_content\u001b[49m\u001b[43m(\u001b[49m\u001b[43mCONTENT_CHUNK_SIZE\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    901\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_content_consumed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    902\u001b[0m \u001b[38;5;66;03m# don't need to release the connection; that's been handled by urllib3\u001b[39;00m\n\u001b[1;32m    903\u001b[0m \u001b[38;5;66;03m# since we exhausted the data.\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/requests/models.py:818\u001b[0m, in \u001b[0;36mResponse.iter_content.<locals>.generate\u001b[0;34m()\u001b[0m\n\u001b[1;32m    816\u001b[0m     \u001b[38;5;28;01myield from\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mstream(chunk_size, decode_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    817\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ProtocolError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 818\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ChunkedEncodingError(e)\n\u001b[1;32m    819\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m DecodeError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    820\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ContentDecodingError(e)\n",
      "\u001b[0;31mChunkedEncodingError\u001b[0m: (\"Connection broken: ConnectionResetError(104, 'Connection reset by peer')\", ConnectionResetError(104, 'Connection reset by peer'))"
     ]
    }
   ],
   "source": [
    "if 'Identifier' in df.columns:\n",
    "    X_pred = df.drop([\"Identifier\"], axis = 1)\n",
    "else:\n",
    "    X_pred = df\n",
    "\n",
    "dict = {k: v.tolist() for k, v in X_pred.iterrows()}\n",
    "\n",
    "brainproteomics_api_url = 'http://127.0.0.1:8000/predict_one_upload'\n",
    "response = requests.get(brainproteomics_api_url, params=dict)\n",
    "\n",
    "prediction = response.json()\n",
    "\n",
    "st.write(prediction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "a8d96fe7-762f-4f6a-b55a-3acbcf51881e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_one_upload(\n",
    "            data # \"Glioma-clinic-TCGA-proteins-test-with-identifier-outcome0.csv\"\n",
    "            ):\n",
    "    \"\"\"\n",
    "    Make a prediction for every row in your dataset.\n",
    "    Input needs to be a csv file with rows = samples and columns = proteins.\n",
    "    The first row has to contain the protein names or any other identifier that will serve as a header.\n",
    "    If your file has a sample identifier column, name this column \"Identifier\"\n",
    "    \"\"\"\n",
    "\n",
    "    X_pred = data\n",
    "\n",
    "\n",
    "    # Preprocess\n",
    "    # Load scaler\n",
    "    scaler = load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "                         file = 'MinMax_20240306-102844.joblib'\n",
    "                        )\n",
    "\n",
    "    X_pred_proc = scaler.transform(X_pred)\n",
    "\n",
    "    # Predict data\n",
    "    model = load_model()\n",
    "\n",
    "    outcome_num = int(model.predict(X_pred_proc)[0])\n",
    "    if outcome_num == 0:\n",
    "        outcome = \"good cancer\"\n",
    "        probability = round(float(model.predict_proba(X_pred_proc)[0][0]), 4)\n",
    "    else:\n",
    "        outcome = \"bad cancer\"\n",
    "        probability = round(float(model.predict_proba(X_pred_proc)[0][1]), 4)\n",
    "\n",
    "    return {\n",
    "                \"Outcome\": outcome,\n",
    "                \"Probability\": probability\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "fe219e16-c18b-457d-bc2a-227989ff6fac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>14-3-3_beta_p</th>\n",
       "      <th>14-3-3_epsilon_p</th>\n",
       "      <th>14-3-3_zeta_p</th>\n",
       "      <th>4E-BP1_p</th>\n",
       "      <th>4E-BP1_pS65_p</th>\n",
       "      <th>4E-BP1_pT37_T46_p</th>\n",
       "      <th>4E-BP1_pT70_p</th>\n",
       "      <th>53BP1_p</th>\n",
       "      <th>A-Raf_pS299_p</th>\n",
       "      <th>ACC1_p</th>\n",
       "      <th>...</th>\n",
       "      <th>p27_p</th>\n",
       "      <th>p27_pT157_p</th>\n",
       "      <th>p27_pT198_p</th>\n",
       "      <th>p38_pT180_Y182_p</th>\n",
       "      <th>p53_p</th>\n",
       "      <th>p62-LCK-ligand_p</th>\n",
       "      <th>p70S6K_p</th>\n",
       "      <th>p70S6K_pT389_p</th>\n",
       "      <th>p90RSK_p</th>\n",
       "      <th>p90RSK_pT359_S363_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.046266</td>\n",
       "      <td>0.010482</td>\n",
       "      <td>0.005867</td>\n",
       "      <td>-0.128982</td>\n",
       "      <td>-0.09789</td>\n",
       "      <td>-0.286694</td>\n",
       "      <td>-0.09536</td>\n",
       "      <td>-0.315572</td>\n",
       "      <td>-0.079862</td>\n",
       "      <td>-0.190765</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282454</td>\n",
       "      <td>0.215619</td>\n",
       "      <td>0.049541</td>\n",
       "      <td>0.394208</td>\n",
       "      <td>-0.173627</td>\n",
       "      <td>-0.088369</td>\n",
       "      <td>-0.160519</td>\n",
       "      <td>0.494404</td>\n",
       "      <td>0.412308</td>\n",
       "      <td>-0.081711</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 174 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   14-3-3_beta_p  14-3-3_epsilon_p  14-3-3_zeta_p  4E-BP1_p  4E-BP1_pS65_p  \\\n",
       "0      -0.046266          0.010482       0.005867 -0.128982       -0.09789   \n",
       "\n",
       "   4E-BP1_pT37_T46_p  4E-BP1_pT70_p   53BP1_p  A-Raf_pS299_p    ACC1_p  ...  \\\n",
       "0          -0.286694       -0.09536 -0.315572      -0.079862 -0.190765  ...   \n",
       "\n",
       "      p27_p  p27_pT157_p  p27_pT198_p  p38_pT180_Y182_p     p53_p  \\\n",
       "0  0.282454     0.215619     0.049541          0.394208 -0.173627   \n",
       "\n",
       "   p62-LCK-ligand_p  p70S6K_p  p70S6K_pT389_p  p90RSK_p  p90RSK_pT359_S363_p  \n",
       "0         -0.088369 -0.160519        0.494404  0.412308            -0.081711  \n",
       "\n",
       "[1 rows x 174 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = X_pred[0:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "9229dc09-624d-40c1-9612-42cf5dcc0751",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator MinMaxScaler from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Outcome': 'bad cancer', 'Probability': 1.0}"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predict_one_upload(X_pred[0:1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "cff19dca-d983-42df-b76a-cbd43bd9ddb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but SGDClassifier was fitted without feature names\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1,\n",
       "       1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1,\n",
       "       1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1])"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = load_model()\n",
    "model.predict(X_pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69970661-9309-41df-950f-d725dc1205cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77803349-cf6f-4c70-9b28-f272713d08cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "a587f9e7-51c3-44ad-9d84-837f3eb1403d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers/',\n",
    "               file = 'scaler.joblib'\n",
    "              ):\n",
    "    \"\"\"\n",
    "    Loads a locally saved and fitted scaler from the given path and file name.\n",
    "    \"\"\"\n",
    "    path_file = f\"{path}/{file}\"\n",
    "\n",
    "    scaler = load(path_file)\n",
    "    return scaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "6ee1477e-b544-4540-a386-c25740c885aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/jfschulz_notebooks'"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "3ba9f38e-e194-4591-b22d-f30b3f7b5953",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-3 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: black;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-3 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-3 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-3 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-3 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-3 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"â–¸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-3 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"â–¾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-3 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-3 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-3 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-3 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-3 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-3 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-3 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-3 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-3 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 1ex;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-3 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-3 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-3 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-3\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>ColumnTransformer(transformers=[(&#x27;pipeline-1&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;ordinalencoder&#x27;,\n",
       "                                                  OrdinalEncoder())]),\n",
       "                                 [&#x27;gender&#x27;]),\n",
       "                                (&#x27;pipeline-2&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;minmaxscaler&#x27;,\n",
       "                                                  MinMaxScaler())]),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x7f7c4d41b130&gt;)])</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-3\" type=\"checkbox\" ><label for=\"sk-estimator-id-3\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;ColumnTransformer<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.4/modules/generated/sklearn.compose.ColumnTransformer.html\">?<span>Documentation for ColumnTransformer</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>ColumnTransformer(transformers=[(&#x27;pipeline-1&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;ordinalencoder&#x27;,\n",
       "                                                  OrdinalEncoder())]),\n",
       "                                 [&#x27;gender&#x27;]),\n",
       "                                (&#x27;pipeline-2&#x27;,\n",
       "                                 Pipeline(steps=[(&#x27;minmaxscaler&#x27;,\n",
       "                                                  MinMaxScaler())]),\n",
       "                                 &lt;sklearn.compose._column_transformer.make_column_selector object at 0x7f7c4d41b130&gt;)])</pre></div> </div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-4\" type=\"checkbox\" ><label for=\"sk-estimator-id-4\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">pipeline-1</label><div class=\"sk-toggleable__content fitted\"><pre>[&#x27;gender&#x27;]</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-5\" type=\"checkbox\" ><label for=\"sk-estimator-id-5\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;OrdinalEncoder<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.4/modules/generated/sklearn.preprocessing.OrdinalEncoder.html\">?<span>Documentation for OrdinalEncoder</span></a></label><div class=\"sk-toggleable__content fitted\"><pre>OrdinalEncoder()</pre></div> </div></div></div></div></div></div></div><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-6\" type=\"checkbox\" ><label for=\"sk-estimator-id-6\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">pipeline-2</label><div class=\"sk-toggleable__content fitted\"><pre>&lt;sklearn.compose._column_transformer.make_column_selector object at 0x7f7c4d41b130&gt;</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-7\" type=\"checkbox\" ><label for=\"sk-estimator-id-7\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;MinMaxScaler<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.4/modules/generated/sklearn.preprocessing.MinMaxScaler.html\">?<span>Documentation for MinMaxScaler</span></a></label><div class=\"sk-toggleable__content fitted\"><pre>MinMaxScaler()</pre></div> </div></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "ColumnTransformer(transformers=[('pipeline-1',\n",
       "                                 Pipeline(steps=[('ordinalencoder',\n",
       "                                                  OrdinalEncoder())]),\n",
       "                                 ['gender']),\n",
       "                                ('pipeline-2',\n",
       "                                 Pipeline(steps=[('minmaxscaler',\n",
       "                                                  MinMaxScaler())]),\n",
       "                                 <sklearn.compose._column_transformer.make_column_selector object at 0x7f7c4d41b130>)])"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "            file = 'scaler.joblib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "74347864-6911-4130-b5ef-068f96f2768d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SVC from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-4 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: black;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-4 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-4 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-4 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"â–¸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"â–¾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-4 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-4 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-4 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-4 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-4 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-4 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-4 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-4 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 1ex;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-4 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-4 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-4\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>SVC(C=10, gamma=&#x27;auto&#x27;)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-8\" type=\"checkbox\" checked><label for=\"sk-estimator-id-8\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;SVC<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.4/modules/generated/sklearn.svm.SVC.html\">?<span>Documentation for SVC</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>SVC(C=10, gamma=&#x27;auto&#x27;)</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "SVC(C=10, gamma='auto')"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "load('/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_models/svc_model.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42cc46ab-e8ed-404f-8a41-3956996d2fa5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2458cac1-a7cd-432f-929b-238661463514",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bccbe22b-2969-4a6d-8de3-ff6007fd10e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55f0389b-08d4-44e5-89d3-868d9c2e4f05",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9afa1312-9a78-4f59-801a-309ac1644343",
   "metadata": {},
   "outputs": [],
   "source": [
    "decoded_str =\"Identifier,14-3-3_beta_p,14-3-3_epsilon_p,14-3-3_zeta_p,4E-BP1_p,4E-BP1_pS65_p,4E-BP1_pT37_T46_p,4E-BP1_pT70_p,53BP1_p,A-Raf_pS299_p,ACC1_p,ACC_pS79_p,ACVRL1_p,AMPK_alpha_p,AMPK_pT172_p,AR_p,ASNS_p,ATM_p,Akt_p,Akt_pS473_p,Akt_pT308_p,Annexin_VII_p,B-Raf_p,BRCA2_p,Bad_pS112_p,Bak_p,Bap1-c-4_p,Bax_p,Bcl-2_p,Bcl-xL_p,Beclin_p,Bid_p,Bim_p,C-Raf_p,C-Raf_pS338_p,CD20_p,CD31_p,CD49b_p,CDK1_p,Caspase-7_cleavedD198_p,Caveolin-1_p,Chk1_p,Chk1_pS345_p,Chk2_p,Chk2_pT68_p,Claudin-7_p,Collagen_VI_p,Cyclin_B1_p,Cyclin_D1_p,Cyclin_E1_p,Cyclin_E2_p,DJ-1_p,Dvl3_p,E-Cadherin_p,EGFR_p,EGFR_pY1068_p,EGFR_pY1173_p,ER-alpha_p,ER-alpha_pS118_p,ERK2_p,ETS-1_p,FASN_p,FOXO3a_p,FOXO3a_pS318_S321_p,Fibronectin_p,FoxM1_p,G6PD_p,GAPDH_p,GATA3_p,GSK3-alpha-beta_p,GSK3-alpha-beta_pS21_S9_p,GSK3_pS9_p,HER2_p,HER2_pY1248_p,HER3_p,HER3_pY1289_p,HSP70_p,Heregulin_p,IGFBP2_p,INPP4B_p,IRS1_p,JNK2_p,Ku80_p,LKB1_p,Lck_p,MAPK_pT202_Y204_p,MEK1_p,MEK1_pS217_S221_p,MIG-6_p,MYH11_p,Mre11_p,N-Cadherin_p,N-Ras_p,NDRG1_pT346_p,NF-kB-p65_pS536_p,NF2_p,Notch1_p,P-Cadherin_p,PAI-1_p,PCNA_p,PDCD4_p,PDK1_p,PDK1_pS241_p,PEA15_p,PEA15_pS116_p,PI3K-p110-alpha_p,PI3K-p85_p,PKC-alpha_p,PKC-alpha_pS657_p,PKC-delta_pS664_p,PKC-pan_BetaII_pS660_p,PR_p,PRAS40_pT246_p,PRDX1_p,PREX1_p,PTEN_p,Paxillin_p,RBM15_p,Rab11_p,Rab25_p,Rad50_p,Rad51_p,Raptor_p,Rb_pS807_S811_p,Rictor_p,Rictor_pT1135_p,S6_p,S6_pS235_S236_p,S6_pS240_S244_p,STAT3_pY705_p,STAT5-alpha_p,Shc_pY317_p,Smad1_p,Smad3_p,Smad4_p,Src_p,Src_pY416_p,Src_pY527_p,Stathmin_p,Syk_p,TAZ_p,TFRC_p,TSC1_p,Transglutaminase_p,Tuberin_p,Tuberin_pT1462_p,VEGFR2_p,XRCC1_p,YAP_p,YAP_pS127_p,YB-1_p,YB-1_pS102_p,beta-Catenin_p,c-Jun_pS73_p,c-Kit_p,c-Met_pY1235_p,c-Myc_p,cIAP_p,eEF2_p,eEF2K_p,eIF4E_p,eIF4G_p,mTOR_p,mTOR_pS2448_p,p21_p,p27_p,p27_pT157_p,p27_pT198_p,p38_pT180_Y182_p,p53_p,p62-LCK-ligand_p,p70S6K_p,p70S6K_pT389_p,p90RSK_p,p90RSK_pT359_S363_p85\\n,0.270059344,0.025498133,0.563840516,-0.168998444,1.062717925,0.139036177,0.111204558,-0.01320412,-0.110908431,0.237457465,0.031126838,0.006204528,0.077297857,0.038855974,-0.209268765,0.27586191,-0.827243768,0.027063721,-0.413822668,0.21295031,0.011577755,0.904158299,0.029589195,-0.161975884,-0.08316436,-0.294166665,-0.178624699,-0.092303592,0.002504112,0.135097572,0.09619223,-0.156119179,-0.076670186,0.062356912,0.033211943,-0.0154909,-0.07985824,0.270102765,0.067704988,-0.799409456,0.006192488,-0.047300732,0.289252684,-0.019121975,0.091497673,0.552778006,-0.180870345,-0.771810267,0.086019922,-0.107921962,0.023191501,-0.292254675,-0.419741443,0.180359664,-0.45779346,0.061767419,-0.006585078,0.013533718,0.3761576,0.451574051,-0.045654254,-0.191386498,-0.161301916,0.363339317,-0.02092265,-0.008160021,-0.228713856,-0.149097685,0.346505118,1.057844255,0.801596648,-0.184302564,-0.070822963,-0.068214017,0.336852072,1.061277972,0.035847954,0.4246895,0.398524747,-0.011429442,0.315622414,-0.404549209,0.296909806,0.165187417,-0.765621283,0.683018866,0.494168111,0.146913094,-0.476038359,-0.113333049,0.114940379,-0.078128625,-0.224089805,-0.479341177,0.178381428,-0.350494183,-0.093606147,0.129189194,0.061753043,-0.550077792,0.34269937,0.463975005,-0.405122392,0.261235219,-0.219388409,-0.008052702,-0.526315135,-0.436082382,-0.320426518,0.290176544,-0.170031679,-0.02441055,-0.319240946,-0.972156515,0.292814001,-0.200702546,-0.36471683,0.076334196,-0.065823494,-0.175150514,0.175834975,-0.002139227,0.855419826,-0.353349182,-0.109112497,-0.252892518,-0.592247541,-0.244121469,-0.53860014,0.091979693,-0.092525245,-0.243962904,0.335473275,0.078724424,0.017584312,0.137519357,0.423657229,0.119194593,-0.312826335,0.139607604,0.041219768,0.100014887,0.3092904,-0.097006742,0.187818637,-0.893805728,0.023924176,-0.065602341,-0.789868599,0.142241429,0.013955578,-0.287002111,-0.278015079,0.431805445,-0.053765795,0.485941082,0.0,0.007169034,-0.347183491,0.014733647,-0.41043721,-0.055778007,0.186858868,0.231578161,-0.115762599,0.092708052,-0.131877322,-0.371950066,0.097358156,-0.377151173,0.102537004,1.762375843,0.045681371,-0.024346193\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b1b3e624-44c1-4496-a8ea-21129af94a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "903e3026-8b3b-4a71-9ff0-dc6c0abfaf49",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e7a4029a-c697-4465-b34d-e81c77864508",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "DataFrame constructor not properly called!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_4284/2055393556.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdumps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdecoded_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    871\u001b[0m                 )\n\u001b[1;32m    872\u001b[0m         \u001b[0;31m# For data is scalar\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    873\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    874\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mcolumns\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 875\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"DataFrame constructor not properly called!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    876\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    877\u001b[0m             \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    878\u001b[0m             \u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: DataFrame constructor not properly called!"
     ]
    }
   ],
   "source": [
    "#json.loads(json.dumps(decoded_str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "593f1b98-32a8-468c-8736-b8ee7181ad4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Identifier</th>\n",
       "      <th>14-3-3_beta_p</th>\n",
       "      <th>14-3-3_epsilon_p</th>\n",
       "      <th>14-3-3_zeta_p</th>\n",
       "      <th>4E-BP1_p</th>\n",
       "      <th>4E-BP1_pS65_p</th>\n",
       "      <th>4E-BP1_pT37_T46_p</th>\n",
       "      <th>4E-BP1_pT70_p</th>\n",
       "      <th>53BP1_p</th>\n",
       "      <th>A-Raf_pS299_p</th>\n",
       "      <th>...</th>\n",
       "      <th>p27_p</th>\n",
       "      <th>p27_pT157_p</th>\n",
       "      <th>p27_pT198_p</th>\n",
       "      <th>p38_pT180_Y182_p</th>\n",
       "      <th>p53_p</th>\n",
       "      <th>p62-LCK-ligand_p</th>\n",
       "      <th>p70S6K_p</th>\n",
       "      <th>p70S6K_pT389_p</th>\n",
       "      <th>p90RSK_p</th>\n",
       "      <th>p90RSK_pT359_S363_p85</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>0.270059344</td>\n",
       "      <td>0.025498133</td>\n",
       "      <td>0.563840516</td>\n",
       "      <td>-0.168998444</td>\n",
       "      <td>1.062717925</td>\n",
       "      <td>0.139036177</td>\n",
       "      <td>0.111204558</td>\n",
       "      <td>-0.01320412</td>\n",
       "      <td>-0.110908431</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.115762599</td>\n",
       "      <td>0.092708052</td>\n",
       "      <td>-0.131877322</td>\n",
       "      <td>-0.371950066</td>\n",
       "      <td>0.097358156</td>\n",
       "      <td>-0.377151173</td>\n",
       "      <td>0.102537004</td>\n",
       "      <td>1.762375843</td>\n",
       "      <td>0.045681371</td>\n",
       "      <td>-0.024346193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 175 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Identifier 14-3-3_beta_p 14-3-3_epsilon_p 14-3-3_zeta_p      4E-BP1_p  \\\n",
       "0              0.270059344      0.025498133   0.563840516  -0.168998444   \n",
       "\n",
       "  4E-BP1_pS65_p 4E-BP1_pT37_T46_p 4E-BP1_pT70_p      53BP1_p A-Raf_pS299_p  \\\n",
       "0   1.062717925       0.139036177   0.111204558  -0.01320412  -0.110908431   \n",
       "\n",
       "   ...         p27_p  p27_pT157_p   p27_pT198_p p38_pT180_Y182_p        p53_p  \\\n",
       "0  ...  -0.115762599  0.092708052  -0.131877322     -0.371950066  0.097358156   \n",
       "\n",
       "  p62-LCK-ligand_p     p70S6K_p p70S6K_pT389_p     p90RSK_p  \\\n",
       "0     -0.377151173  0.102537004    1.762375843  0.045681371   \n",
       "\n",
       "  p90RSK_pT359_S363_p85  \n",
       "0          -0.024346193  \n",
       "\n",
       "[1 rows x 175 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "#pd.read_csv(StringIO(decoded_str))\n",
    "\n",
    "rows = decoded_str.split('\\n')\n",
    "\n",
    "# Split each row into columns\n",
    "data = [row.split(',') for row in rows]\n",
    "\n",
    "# Convert the data into a DataFrame\n",
    "df = pd.DataFrame(data[1:], columns=data[0])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d029c745-784b-4962-9e86-fa7909482f8d",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'to_json'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdecoded_str\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_json\u001b[49m()\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'to_json'"
     ]
    }
   ],
   "source": [
    ".to_json()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1950005c-e500-4d43-853e-b27d0e0af192",
   "metadata": {},
   "source": [
    "## Perform prediction for several samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "855f53b7-4577-4dea-b69a-8193c3009b31",
   "metadata": {},
   "source": [
    "### Functions needed to perform prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7567b0b4-38c1-4bf4-88c0-4f3a74326e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_data (df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Clean raw data:\n",
    "    - remove duplicates\n",
    "    - remove NaNs\n",
    "    \"\"\"\n",
    "\n",
    "    df = df.drop_duplicates()\n",
    "    df = df.dropna(how='any', axis=0)\n",
    "\n",
    "    #clinical_data= ['Case', 'years_to_birth', 'gender', 'histological_type', 'race', 'ethnicity', 'radiation_therapy', 'Grade', 'Mutation.Count', 'Percent.aneuploidy', 'IDH.status', 'outcome' ]\n",
    "\n",
    "    print(\"âœ… Data cleaned\")\n",
    "    return df\n",
    "\n",
    "# Load scaler\n",
    "def load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "               file = 'scaler.gz'\n",
    "              ):\n",
    "    \"\"\"\n",
    "    Loads a locally saved and fitted scaler from the given path and file name.\n",
    "    \"\"\"\n",
    "    path_file = f\"{path}/{file}\"\n",
    "\n",
    "    scaler = load(path_file)\n",
    "    return scaler\n",
    "\n",
    "\n",
    "def preprocess_input(data) -> np.array:\n",
    "    \"\"\"\n",
    "    - process data (proteins, age and gender)\n",
    "    - load fitted scaler and pipline for preprocessing\n",
    "    - return processed data\n",
    "    \"\"\"\n",
    "    print(\" Preprocess input proteins, age and gender... \")\n",
    "\n",
    "    # Clean data\n",
    "    data_clean= clean_data(data)\n",
    "\n",
    "    # define X\n",
    "    X = data_clean.drop(['Case', 'histological_type', 'race', 'ethnicity', 'radiation_therapy', 'Grade', 'Mutation.Count', 'Percent.aneuploidy', 'IDH.status'], axis = 1)\n",
    "\n",
    "    preproc_scaler = load_scaler(path = '/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_scalers',\n",
    "                                 file = 'scaler.joblib'\n",
    "                                )\n",
    "    # preprocess X_train, X_test and X_val\n",
    "    X_pred= preproc_scaler.transform(X)\n",
    "\n",
    "    return X_pred\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d7a8655-4d72-482f-a004-1326ce2b2940",
   "metadata": {},
   "source": [
    "## do most of the table wrangling in the api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "a3d4cec1-58c4-4b41-9a7b-a97f9d0a4354",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Preprocess input proteins, age and gender... \n",
      "âœ… Data cleaned\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jana/.pyenv/versions/3.10.6/envs/brain_proteomics/lib/python3.10/site-packages/sklearn/base.py:376: InconsistentVersionWarning: Trying to unpickle estimator SGDClassifier from version 1.4.0 when using version 1.4.1.post1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "df_upload = pd.read_csv(\"/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/raw_data/Glioma-clinic-TCGA-proteins-all-cols-test-with-identifier.csv\")\n",
    "\n",
    "df = df_upload.drop(columns=\"Identifier\", axis = 1)\n",
    "\n",
    "# Preprocess\n",
    "X_pred = preprocess_input(df)\n",
    "\n",
    "# Predict data\n",
    "model = load_model(path ='/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/saved_models',\n",
    "                    file = 'sgd_model.pkl')\n",
    "\n",
    "outcome = pd.DataFrame(model.predict(X_pred), columns=[\"Outcome\"], dtype = int)\n",
    "prob = pd.DataFrame(model.predict_proba(X_pred), columns=[\"Probability_0\", \"Probability_1\"], dtype = float)\n",
    "\n",
    "# Merge results into dataframe\n",
    "result = pd.merge(prob,outcome, left_index=True, right_index=True)\n",
    "#result_dict = result.to_dict('series')\n",
    "\n",
    "#prediction = {k: v.tolist() for k, v in result.iterrows()}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c8dc8a-aa50-452e-85e6-75c2d20e0692",
   "metadata": {},
   "source": [
    "### Create Plots for results table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "1fcb99a1-0669-4abf-9dc8-423cb0bbcd58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index</th>\n",
       "      <th>Prediction</th>\n",
       "      <th>Probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>Astrocytoma</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>259</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>237</td>\n",
       "      <td>Oligodendroglioma</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>Astrocytoma</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>180</td>\n",
       "      <td>Astrocytoma</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Index         Prediction Probability\n",
       "0    12        Astrocytoma         1.0\n",
       "1   259  Oligodendroglioma         1.0\n",
       "2   237  Oligodendroglioma         1.0\n",
       "3    19        Astrocytoma         1.0\n",
       "4   180        Astrocytoma         1.0"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import plotly.express as px\n",
    "result_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "8bb2a8a3-92fe-4577-9a11-b54b484eb8c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this will be in the API\n",
    "# transform dictionary output into dataframe and bring data frame to final display shape\n",
    "result_df[\"Prediction\"] = result[\"Outcome\"].apply(lambda x: \"Oligodendroglioma\" if x == 0 else \"Astrocytoma\")\n",
    "\n",
    "result_df[\"Probability\"] = np.where(result[\"Outcome\"] == 0, \n",
    "                                result[\"Probability_0\"], \n",
    "                                result[\"Probability_1\"])\n",
    "\n",
    "result_df = df_upload[[\"Identifier\"]].merge(result_df, left_index=True, right_index=True) # add idenfitier column from uploaded dataframe back to output\n",
    "prediction = {k: v.tolist() for k, v in result_df.iterrows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "9cf7dcdd-e635-4c44-8bb6-73650fb26c87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this will be in streamlit\n",
    "result_df = pd.DataFrame(prediction).transpose()\n",
    "result_df = result_df.rename(columns={0: \"Index\", \n",
    "                                      1: \"Prediction\", \n",
    "                                      2: \"Probability\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "22919e11-6b4c-46c5-8877-ba38dadd518d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Astrocytoma', 'Oligodendroglioma'], dtype=object)"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import plotly.express as px\n",
    "fig = px.box(result_df, x='Prediction', y='Probability', color='Prediction')\n",
    "             #title='Boxplots of the differentially expressed proteins',\n",
    "             #category_orders={'outcome': ['Oligodendroglioma', 'Astrocytoma']}\n",
    "            #)\n",
    "#fig.show()\n",
    "result_df[\"Prediction\"].value_counts().values\n",
    "result_df[\"Prediction\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "149632f3-d549-4b9d-83cf-c8ef69bc7e92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df = pd.DataFrame\n",
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "746d8f9f-910f-4379-91cc-3307f369f079",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "8ad76c35-e56d-4c3d-b62c-570dfbb4c3c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jana/code/Klara-haas/brain_proteomics_project/brain_proteomics/api/jfschulz_notebooks'"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
